<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Elsa B.F.">

<title>Rstudio Scripts</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-1d4f84655f446305ac42a8f1abcf7405.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="quarto-light">

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#rstudio-basics" id="toc-rstudio-basics" class="nav-link active" data-scroll-target="#rstudio-basics">Rstudio basics</a>
  <ul class="collapse">
  <li><a href="#recomanacions-inicials" id="toc-recomanacions-inicials" class="nav-link" data-scroll-target="#recomanacions-inicials">Recomanacions inicials</a></li>
  <li><a href="#bàsics" id="toc-bàsics" class="nav-link" data-scroll-target="#bàsics">Bàsics</a>
  <ul class="collapse">
  <li><a href="#installació" id="toc-installació" class="nav-link" data-scroll-target="#installació">Instal·lació</a></li>
  <li><a href="#zona-de-treball" id="toc-zona-de-treball" class="nav-link" data-scroll-target="#zona-de-treball">Zona de treball</a></li>
  <li><a href="#script" id="toc-script" class="nav-link" data-scroll-target="#script">Script</a></li>
  <li><a href="#installar-i-carregar-paquets" id="toc-installar-i-carregar-paquets" class="nav-link" data-scroll-target="#installar-i-carregar-paquets">Instal·lar i carregar paquets</a></li>
  <li><a href="#importar-dades" id="toc-importar-dades" class="nav-link" data-scroll-target="#importar-dades">Importar dades</a></li>
  <li><a href="#descarregar-dades-i-figures" id="toc-descarregar-dades-i-figures" class="nav-link" data-scroll-target="#descarregar-dades-i-figures">Descarregar dades i figures</a></li>
  <li><a href="#altres-bàsics" id="toc-altres-bàsics" class="nav-link" data-scroll-target="#altres-bàsics">Altres bàsics</a></li>
  <li><a href="#programació-en-r" id="toc-programació-en-r" class="nav-link" data-scroll-target="#programació-en-r">Programació en R</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#anàlisi-de-dades" id="toc-anàlisi-de-dades" class="nav-link" data-scroll-target="#anàlisi-de-dades">Anàlisi de dades</a>
  <ul class="collapse">
  <li><a href="#transformar-el-dataset" id="toc-transformar-el-dataset" class="nav-link" data-scroll-target="#transformar-el-dataset">Transformar el dataset</a>
  <ul class="collapse">
  <li><a href="#tidyverse" id="toc-tidyverse" class="nav-link" data-scroll-target="#tidyverse">Tidyverse</a></li>
  <li><a href="#crear-nous-dataset" id="toc-crear-nous-dataset" class="nav-link" data-scroll-target="#crear-nous-dataset">Crear nous dataset</a></li>
  <li><a href="#wide-format-vs-long-format" id="toc-wide-format-vs-long-format" class="nav-link" data-scroll-target="#wide-format-vs-long-format">Wide format vs Long format</a></li>
  </ul></li>
  <li><a href="#estadística-descriptiva" id="toc-estadística-descriptiva" class="nav-link" data-scroll-target="#estadística-descriptiva">Estadística descriptiva</a></li>
  <li><a href="#inferència-estadística" id="toc-inferència-estadística" class="nav-link" data-scroll-target="#inferència-estadística">Inferència estadística</a>
  <ul class="collapse">
  <li><a href="#normalitat" id="toc-normalitat" class="nav-link" data-scroll-target="#normalitat">Normalitat</a></li>
  <li><a href="#homogeneïtat-de-les-variàncies" id="toc-homogeneïtat-de-les-variàncies" class="nav-link" data-scroll-target="#homogeneïtat-de-les-variàncies">Homogeneïtat de les variàncies</a></li>
  <li><a href="#independent-samples-t-test" id="toc-independent-samples-t-test" class="nav-link" data-scroll-target="#independent-samples-t-test">Independent samples T-test</a></li>
  <li><a href="#one-way-anova" id="toc-one-way-anova" class="nav-link" data-scroll-target="#one-way-anova">One-way ANOVA</a></li>
  <li><a href="#u-mann-whitney" id="toc-u-mann-whitney" class="nav-link" data-scroll-target="#u-mann-whitney">U-Mann Whitney</a></li>
  <li><a href="#kruskall-wallis" id="toc-kruskall-wallis" class="nav-link" data-scroll-target="#kruskall-wallis">Kruskall-Wallis</a></li>
  <li><a href="#art-anova" id="toc-art-anova" class="nav-link" data-scroll-target="#art-anova">Art ANOVA</a></li>
  </ul></li>
  <li><a href="#correlacions" id="toc-correlacions" class="nav-link" data-scroll-target="#correlacions">Correlacions</a></li>
  </ul></li>
  <li><a href="#gràfics" id="toc-gràfics" class="nav-link" data-scroll-target="#gràfics">Gràfics</a>
  <ul class="collapse">
  <li><a href="#pcr" id="toc-pcr" class="nav-link" data-scroll-target="#pcr">PCR</a></li>
  <li><a href="#heatmap-log2fc" id="toc-heatmap-log2fc" class="nav-link" data-scroll-target="#heatmap-log2fc">Heatmap log<sub>2</sub>(FC)</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Rstudio Scripts</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Elsa B.F. </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="rstudio-basics" class="level1">
<h1>Rstudio basics</h1>
<p>Bon dia i benvinguts al codi més interessant per a nosaltres en Rstudio. He creat aquesta web per compartir els meus coneixements adquirits al respecte i per anar-la actualitzant a mesura que n’apregui més. M’agradaria comentar que no s’ha fet desde la perfecció lingüística, sinó desde la rapidesa. Espero que no m’ho tingueu en compte.</p>
<p>Tant de bo us sigui útil, Elsa.</p>
<section id="recomanacions-inicials" class="level2">
<h2 class="anchored" data-anchor-id="recomanacions-inicials">Recomanacions inicials</h2>
<p>Si sou com jo i vau aprendre R abans de que existís la intel·ligència artificial, probablement ho recordareu amb por, però això ja no és així.</p>
<p>Per començar recomano sempre tenir una idea de què necessites i vols fer. Seguidament buscar a internet alguna persona que ho expliqui, ja sigui una web o un video i seguir les seves instruccions. Finalment, anar provant el codi amb les teves dades i quan trobis algun error, o vulguis canviar alguna característica, explicar-li a alguna intel·ligència artificial que tens, on vols arribar i quins problemes t’estàs trobant.</p>
<p>No recomano (de moment) anar directament a la intel·ligència artificial.</p>
<p>Us deixo algunes webs que a mi em van anar bé quan vaig tornar aprendre de veritat a treballar amb Rstudio:</p>
<ul>
<li><p>Cursos de Rstudio: <a href="https://education.rstudio.com/learn/beginner/" class="uri">https://education.rstudio.com/learn/beginner/</a></p></li>
<li><p>Basics: <a href="https://rladiessydney.org/courses/01-basicbasics-0" class="uri">https://rladiessydney.org/courses/01-basicbasics-0</a></p></li>
<li><p>Estadística: <a href="https://learningstatisticswithr.com/book/" class="uri">https://learningstatisticswithr.com/book/</a></p></li>
</ul>
<p>I alguns que encara no he mirat però poden ser interessants</p>
<ul>
<li>R for data science: <a href="https://r4ds.hadley.nz/" class="uri">https://r4ds.hadley.nz/</a></li>
<li>Ggplot: <a href="https://ggplot2-book.org/" class="uri">https://ggplot2-book.org/</a></li>
<li>Graph cookbook: <a href="https://r-graphics.org/index.html" class="uri">https://r-graphics.org/index.html</a></li>
<li>Data visualization: <a href="https://socviz.co/" class="uri">https://socviz.co/</a></li>
</ul>
</section>
<section id="bàsics" class="level2">
<h2 class="anchored" data-anchor-id="bàsics">Bàsics</h2>
<section id="installació" class="level3">
<h3 class="anchored" data-anchor-id="installació">Instal·lació</h3>
<p>Per començar ens haurem de descarregar:</p>
<ul>
<li><p>R (<a href="https://ftp.cixug.es/CRAN/index.html" class="uri">https://ftp.cixug.es/CRAN/index.html</a>)</p></li>
<li><p>Rstudio (<a href="https://posit.co/download/rstudio-desktop/" class="uri">https://posit.co/download/rstudio-desktop/</a>)</p></li>
</ul>
<p>Un cop ho tingueu instal·lat ja podem començar.</p>
</section>
<section id="zona-de-treball" class="level3">
<h3 class="anchored" data-anchor-id="zona-de-treball">Zona de treball</h3>
<p>Tenim dues maneres de treballar, creant un projecte o agafant una carpeta qualsevol i treballant en un directori de treball.</p>
<p>La més senzilla és creant un directori de treball (working directory). Allà hem de tenir el nostre document .csv o .xlsx amb les dades. També serà la carpeta on ens guardarà imatges o documents en cas que ho demanem mitjançant codi.</p>
<p>Es pot indicar de la següent manera:</p>
<p><img src="images/Captura de pantalla 2026-01-16 105625-01.png" class="img-fluid" width="430"></p>
<p>També es pot indicar mitjançant codi (jo ho prefereixo):</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"C:/pathway que ha de seguir per arribar a la carpeta d'interés"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Per altre banda es pot crear un projecte (millor si ho vols tenir tot més organitzat):</p>
<p><img src="images/Captura de pantalla 2026-01-16 110310.png" class="img-fluid" width="317"></p>
<p>En aquest cas pots crear diferents carpetes a dins del projecte directament desde la finestra de “Files” a l’extrem dret inferior. Recomano crear-ne tres mínim anomenats:</p>
<ul>
<li><p>“data”: col·locar el fitxer amb les dades</p></li>
<li><p>“scripts”: on guardaràs els scripts</p></li>
<li><p>“output”: on gruardar els gràfics</p></li>
</ul>
<p>En el cas que feu análisis en més d’un dataset, recomano crear carpetes per a cada un, per exemple “PCR” i dins d’aquesta les tres que acabo de dir.</p>
</section>
<section id="script" class="level3">
<h3 class="anchored" data-anchor-id="script">Script</h3>
<p>L’script son les instruccions que aniràs escrivint. Alguns trucs que he anat descobrint són…</p>
<p>Si escrius una “ordre”, apareix en negre i s’executarà al tenir en cursor a la línia (o seleccionar-ne una o varies) i posar Run. Una altre manera que a mi m’agrada més és amb el teclat: <strong>Ctrl + Enter</strong></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(data) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Per posar un comentari poses un <code>#</code> i tot el que estigui darrere no s’executarà</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#en aquest cas no s'executarà res</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(data) <span class="co">#en aquest altre cas executarà la ordre i tindràs el comentari</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co">#View(data) aquest cas és incorrecte si vols executar l'ordre ja que no ho farà</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Una altre característica xula és que a dins l’script pots anar creant un índex, que apareix a l’outline. Per activar-ho simplement has de colocar mínim 4 <code>----</code> al final del comentari. Al posar diferent número de <code>#</code> crearas diferent subnivells.</p>
<p>Això també fa que aparegui una pestanyeta a l’esquerra del títol i permetrà amagar tota aquella secció, fent la revisió i lectura de l’script sencer més còmode.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Títol 1 ----</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Subtítol----</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="do">### Sub-subtítol ----</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="do">#### encara més subtitol ----</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Títol 2 ----</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p><img src="images/Captura de pantalla 2026-01-16 134038.png" class="img-fluid" width="194"></p>
</section>
<section id="installar-i-carregar-paquets" class="level3">
<h3 class="anchored" data-anchor-id="installar-i-carregar-paquets">Instal·lar i carregar paquets</h3>
<p>R té funcions bàsiques, però per fer-te la vida més fàcil molta gent ha creat funcions que et pots descarregar en paquets. El primer cop que l’utilitzis has d’instal·lar-lo i carregar-lo i els següents només carregar-lo.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"nom paquet"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(nom paquet)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Els que jo sempre utilitzo són:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co">#conté 9 paquets d'edició i visualització </span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here) <span class="co">#al treballar en un projecte</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl) <span class="co">#per importar dades que estan en un fitxer d'excel</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor) <span class="co">#ajudar a netejar les dades</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(skimr) <span class="co">#per obtenir un resum de les dades complert</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(broom) <span class="co">#poder fer servir funcions com tidy</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstatix) <span class="co">#funcions d'estadística</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="importar-dades" class="level3">
<h3 class="anchored" data-anchor-id="importar-dades">Importar dades</h3>
<p>Amb la zona de treball preparada i els paquets instal·lats i carregats ja podem començar a treballar, i el primer serà importar les dades.</p>
<p>Desde un working directory:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>raw_data<span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"nom excel.xlsx"</span>, <span class="at">sheet =</span> <span class="st">"nom full"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>En un projecte amb el paquet <code>library(here)</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"nom excel.xlsx"</span>),  <span class="at">sheet =</span> <span class="st">"nom full"</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#en el cas de estar en una subcarpeta</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">here</span>(<span class="st">"nom subcarpeta"</span>, <span class="st">"data"</span>, <span class="st">"nom excel.xlsx"</span>),  <span class="at">sheet =</span> <span class="st">"nom full"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Seguidament comprovem que s’han importat bé.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(raw_data) <span class="co">#Mirar tota la taula</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(raw_data) <span class="co">#imprimeix només el principi de les dades</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(raw_data) <span class="co">#obtenir estadístics descriptius i característiques de les variables</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Si tot està correcte ja podem començar amb el que vulguem.</p>
</section>
<section id="descarregar-dades-i-figures" class="level3">
<h3 class="anchored" data-anchor-id="descarregar-dades-i-figures">Descarregar dades i figures</h3>
<p>Abans de passar a l’anàlisi de dades us indico com descarregar-vos arxius que hagueu generat:</p>
<ul>
<li>Dataframes</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(writexl)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write_xlsx</span>(nom_dataframe, <span class="st">"nom fitxer.xlsx"</span>)  <span class="co">#amb working directory</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write_xlsx</span>(nom_dataframe, <span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"nom fitxer.csv"</span>))   <span class="co">#amb projecte</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Figures amb <code>ggplot</code> (paquet inclós dins de tidyverse )</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">here</span>(<span class="st">"output"</span>,<span class="st">"nom fitxer.png"</span>), <span class="at">width =</span> <span class="dv">12</span>, <span class="at">height =</span> <span class="dv">3</span>, <span class="at">dpi =</span> <span class="dv">300</span>) </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co">#pots canviar la mida i la resolució (dpi)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="altres-bàsics" class="level3">
<h3 class="anchored" data-anchor-id="altres-bàsics">Altres bàsics</h3>
<p>Finalment, escriuré funcions molt bàsiques que segurament ja conegueu però per tenir-les recollides.</p>
<p>R és una calculadora amb memòria per tant es poden fer operacions bàsiques, suma, resta…</p>
<p>Generar objectes <code>obj &lt;- x</code> (Shortcut per a posar <code>&lt;-</code>: alt + -)</p>
<p>Definir vectors <code>vec &lt;- c(x, y, z)</code></p>
<p>Definir matrius <code>mat &lt;- matrix(data=c(x, y, z), nrow=X, ncol=Y)</code></p>
<p>Accedir a elements del dataframe:</p>
<ul>
<li><p>Columna pel nom <code>data$nomcolumna</code></p></li>
<li><p>Element per coordenades <code>data[fila, columna]</code></p></li>
<li><p>Rang dades <code>data[x:y,] #(buit=tot)</code></p></li>
</ul>
<p>Funcions interessants primera vista a les dades (i comprovar que tot bé)</p>
<p><code>dim(data)</code> , <code>str(data)</code> , <code>glimpse(data)</code> , <code>head(data) / tail(data)</code> , <code>View(data)</code> , <code>glimpse(data)</code></p>
<p>Netejar dades Netejar dades: canviar tots els caràcters dels noms de les columnes</p>
<p><code>data &lt;- select_all(data, tolower/toupper)</code></p>
<p><code>data &lt;- rename(data, newname=oldname)</code></p>
<p>Reordenar columnes o triar quines vols en un nou dataframe</p>
<p><code>select(data, col1, col2, col3, everything())</code></p>
<p>En el cas de resultats obtinguts, per exemple result.test, la majoria són elements tipus llista.</p>
<ul>
<li><p>Accedir element concret: <code>Nomllista$nomobjecte (result.test$p.value)</code></p></li>
<li><p>Veure tots elements <code>str(nomllista)</code></p></li>
</ul>
<p>Separa o ajuntar columnes (i mantenir inicial): <code>separate /unite (remove=F)</code></p>
<p>Canviar o afegir columnes amb operacions <code>mutate</code></p>
<p>… entre moltes d’altres que ja aniré explicant.</p>
</section>
<section id="programació-en-r" class="level3">
<h3 class="anchored" data-anchor-id="programació-en-r">Programació en R</h3>
<p>Realment R és un llenguatge de programació, per tant es poden fer altres coses molt interessants com loops. Aquí ja no entraré però si estàs interessat que sàpiques que existeixen i a vegades el ChatGPT en crea, per tant que els sàpigues interpretar si t’apareixen.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Loops</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> ( CONDITION ) {</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  STATEMENT1</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  STATEMENT2</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  ETC</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>} <span class="co">#mentre la condició es compleixi anirà fent el que tingui a dins i repetirà fins que es deixi de complir</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ( VAR <span class="cf">in</span> VECTOR ) {</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  STATEMENT1</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  STATEMENT2</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  ETC</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>} <span class="co">#nombre finit, per cada variable al vector farà el que li demanis fins que acabi amb totes. </span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> ( CONDITION ) { <span class="co">#si aquesta condició es compleix farà els statement 1 i 2</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  STATEMENT1</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  STATEMENT2</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  ETC</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> { <span class="co">#si no es compleix l'altre fer aquesta</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  STATEMENT3</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>  STATEMENT4</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  ETC</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>} </span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>FNAME <span class="ot">&lt;-</span> <span class="cf">function</span> ( ARG1, ARG2, ETC ) { <span class="co">#funcions</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>  STATEMENT1</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>  STATEMENT2</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>  ETC</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>( VALUE )</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
</section>
<section id="anàlisi-de-dades" class="level1">
<h1>Anàlisi de dades</h1>
<p>Perfecte, doncs ara que ja sabem què és un script i com carregar dades començarem a tractar-les.</p>
<p>El meu tip és que tot i que per a fer un anàlisi no necessitis certes variables, tu les col·loquis totes al teu full de càlcul. En el meu cas jo estic treballant amb rates que reben una dieta materna i treballo tant amb les mares com amb la descendència. Per tant faré les següents columnes sempre:</p>
<ul>
<li><p><strong>rat</strong>: identificador de la rata</p></li>
<li><p><strong>generation</strong>: a quina generació pertanyen (“M”, “Cd1”, “Cd21”, “CA”)</p></li>
<li><p><strong>sex</strong>: el sexe de la rata (“M”, “F”, “ns”)</p></li>
<li><p><strong>group</strong>: a quin grup de dieta pertanyen (“REF”, “P”, “G”, “S”, “PGS”)</p></li>
<li><p><strong>variables d’interés</strong>: columnes diferents amb títol de cada variable i els valors que ha obtingut cada rata.</p>
<ul>
<li><p>PCR: una columna per cada gen</p></li>
<li><p>Lipidòmica: una columna per cada lípid</p></li>
<li><p>Beads: una columna per cada Ig</p></li>
<li><p>i un llarg etcètera</p></li>
</ul></li>
</ul>
<p>Si un dia jo només tinc rates que son de la generació “M”, igualment fer una columna i indicar-ho, ja que un cop a dins de R, podem treure columnes si no ens interessen, o ajuntar dos datasets diferents on llavors potser sí que és interessant tenir aquesta informació.</p>
<p>El que si que no recomano és tenir dues columnes amb la mateixa informació, com per exemple una columna on el grup estigui codificat amb lletres i una altre amb números. Si en algun moment ho necessitem, simplement mutarem la columna existent i ja està.</p>
<p>Una altre recomanació és fer els noms de columnes, i les variables a dins seu, el més simple possible. Això és degut a que després, quan creem el codi, segurament ho haurem d’escriure EXACTAMENT igual, per tant com més simple menys maldecaps. Si esteu fent servir algun dataset que no teniu preparat per això no us preocupeu, hi ha maneres de fer-lo més simple dins d’R.</p>
<ul>
<li><p><em>grup</em> enlloc de <em>Grup</em></p></li>
<li><p><em>M</em> enlloc de <em>Mascles</em></p></li>
</ul>
<p>A més a més el meu dataset és complicat ja que no tinc el mateix nombre de grups per cada generació. Aquí teniu una taula resum que serà útil per entendre els anàlisis estadístics posteriors.</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Generació</th>
<th>Grups</th>
<th>Sexe</th>
<th>Rates per grup</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>M</td>
<td>REF, P, G, S, PGS</td>
<td>F</td>
<td>6</td>
</tr>
<tr class="even">
<td>Cd1</td>
<td>REF, P, G, PGS</td>
<td>M, F</td>
<td>10</td>
</tr>
<tr class="odd">
<td>Cd21</td>
<td>REF, P, G, S, PGS</td>
<td>M, F</td>
<td>10</td>
</tr>
<tr class="even">
<td>CA</td>
<td>REF, PGS</td>
<td>M, F</td>
<td>10</td>
</tr>
</tbody>
</table>
<p>Per a que no hi hagi problemes de filtració de dades li he demanat a ChatGPT que em crei un dataset d’aquestes característiques, així que totes les dades que faré servir i els resultats són inventats.</p>
<p>Això si, jo ja he fet un pre-processat de les dades amb l’excel en el que he tret els outliers de cada grup (per tant alguna cel·la estarà buida i tindré <code>NA</code>) i els valors que introdueixo són expressió relativa respecte el grup REF de la seva generació, en %. Si no ho teniu fet es podria fer tot dins de l’R però jo per començar no m’he complicat tant.</p>
<p>L’excel llueix així:</p>
<p><img src="images/Captura de pantalla 2026-01-16 155245.png" class="img-fluid" width="459"></p>
<p>Importem les dades i comprovem que estan bé i els grups que existeixen corresponen al que hem indicat:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">here</span>(<span class="st">"Basic"</span>, <span class="st">"data"</span>, <span class="st">"PCR.xlsx"</span>),  <span class="at">sheet =</span> <span class="st">"relative"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(raw_data)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(skimr)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(raw_data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">raw_data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">140</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">8</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">character</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 19%">
<col style="width: 13%">
<col style="width: 19%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">min</th>
<th style="text-align: right;">max</th>
<th style="text-align: right;">empty</th>
<th style="text-align: right;">n_unique</th>
<th style="text-align: right;">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">generation</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">group</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">sex</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">rat</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">122</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 11%">
<col style="width: 16%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 6%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gene_1</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">115.60</td>
<td style="text-align: right;">28.32</td>
<td style="text-align: right;">61.3</td>
<td style="text-align: right;">93.70</td>
<td style="text-align: right;">106.7</td>
<td style="text-align: right;">139.25</td>
<td style="text-align: right;">170.7</td>
<td style="text-align: left;">▂▇▃▃▃</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_2</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">119.72</td>
<td style="text-align: right;">30.46</td>
<td style="text-align: right;">67.2</td>
<td style="text-align: right;">97.00</td>
<td style="text-align: right;">111.4</td>
<td style="text-align: right;">141.20</td>
<td style="text-align: right;">192.4</td>
<td style="text-align: left;">▅▇▅▃▂</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gene_3</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">0.93</td>
<td style="text-align: right;">119.48</td>
<td style="text-align: right;">29.84</td>
<td style="text-align: right;">50.7</td>
<td style="text-align: right;">97.30</td>
<td style="text-align: right;">109.0</td>
<td style="text-align: right;">146.48</td>
<td style="text-align: right;">182.1</td>
<td style="text-align: left;">▁▇▅▅▃</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_4</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">119.15</td>
<td style="text-align: right;">29.24</td>
<td style="text-align: right;">67.0</td>
<td style="text-align: right;">96.05</td>
<td style="text-align: right;">110.2</td>
<td style="text-align: right;">143.40</td>
<td style="text-align: right;">182.6</td>
<td style="text-align: left;">▂▇▃▅▂</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Com veiem, ens ha fet un resum general de les dades i variables, totes com a conjunt.</p>
<p>Anem a teure’ls-hi suc.</p>
<section id="transformar-el-dataset" class="level2">
<h2 class="anchored" data-anchor-id="transformar-el-dataset">Transformar el dataset</h2>
<section id="tidyverse" class="level3">
<h3 class="anchored" data-anchor-id="tidyverse">Tidyverse</h3>
<p>Per tal de fer-nos la vida més fàcil, altre gent que treballa amb R ha creat el paquet <code>tidyverse</code> el qual, com ja he dit, conté els paquets més importants per a transformar dades i graficar-les. També hi ha gent molt generosa que ha fet CheatSheets dels paquets més útils, que us recomano imprimir i tenir sempre a mà: <a href="https://posit.co/resources/cheatsheets/" class="uri">https://posit.co/resources/cheatsheets/</a></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.1     ✔ stringr   1.5.2
✔ ggplot2   4.0.0     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
S'està adjuntant el paquet: 'janitor'

Els següents objectes estan emmascarats des de 'package:stats':

    chisq.test, fisher.test</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(broom) </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstatix)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>
S'està adjuntant el paquet: 'rstatix'

L'objecte següent està emmascarat per 'package:janitor':

    make_clean_names

L'objecte següent està emmascarat per 'package:stats':

    filter</code></pre>
</div>
</div>
<p>Una de les característiques més interessants és la pipeline <code>%&gt;%</code> (o ara més nou de R <code>|&gt;</code> ). Aquesta funció del paquet <code>dyplr</code> permet enllaçar moltes accions i que vagin succeint de manera seqüencial en un dataframe.</p>
<p>Primer s’indica el dataframe i després comeces a fer accions. Per tant dins d’aquestes accions ja no cal que indiquis quin dataset ha d’agafar.</p>
<p>Per altra banda pots fer certes accions a partir de condicions que es troben en un altre dataset (ja ho veurem més endavant).</p>
</section>
<section id="crear-nous-dataset" class="level3">
<h3 class="anchored" data-anchor-id="crear-nous-dataset">Crear nous dataset</h3>
<p>El següent pas un cop hem importat el dataset és determinar quin tipus de variable és cada una, i a mi m’agrada crear un nou dataset amb les modificacions i deixar el <code>raw_data</code> intacte per si en algun moment necessitem alguna dada o fem modificacions irreversibles sense voler no haver de tornar a importar.</p>
<p>Si volem grardar el nou dataset és important posar davant de l’acció data <code>nom_dataset &lt;-</code> d’aquesta manera crearem un dataset amb el nom que triem amb les accions que hem indicat.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> raw_data <span class="sc">%&gt;%</span> <span class="co">#farà les accions sobre el dataset anomenat raw_data</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(group, sex, rat, generation), as.factor)) <span class="co">#definir totes les variables que son factors</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">140</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">8</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">factor</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: factor</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 11%">
<col style="width: 16%">
<col style="width: 9%">
<col style="width: 10%">
<col style="width: 36%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: left;">ordered</th>
<th style="text-align: right;">n_unique</th>
<th style="text-align: left;">top_counts</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">generation</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">Cd2: 50, Cd1: 40, M: 30, CA: 20</td>
</tr>
<tr class="even">
<td style="text-align: left;">group</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">PGS: 36, REF: 36, G: 26, P: 26</td>
</tr>
<tr class="odd">
<td style="text-align: left;">sex</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F: 85, M: 55</td>
</tr>
<tr class="even">
<td style="text-align: left;">rat</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">122</td>
<td style="text-align: left;">G1.: 2, G3.: 2, G8.: 2, P2.: 2</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 11%">
<col style="width: 16%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 6%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gene_1</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">115.60</td>
<td style="text-align: right;">28.32</td>
<td style="text-align: right;">61.3</td>
<td style="text-align: right;">93.70</td>
<td style="text-align: right;">106.7</td>
<td style="text-align: right;">139.25</td>
<td style="text-align: right;">170.7</td>
<td style="text-align: left;">▂▇▃▃▃</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_2</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">119.72</td>
<td style="text-align: right;">30.46</td>
<td style="text-align: right;">67.2</td>
<td style="text-align: right;">97.00</td>
<td style="text-align: right;">111.4</td>
<td style="text-align: right;">141.20</td>
<td style="text-align: right;">192.4</td>
<td style="text-align: left;">▅▇▅▃▂</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gene_3</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">0.93</td>
<td style="text-align: right;">119.48</td>
<td style="text-align: right;">29.84</td>
<td style="text-align: right;">50.7</td>
<td style="text-align: right;">97.30</td>
<td style="text-align: right;">109.0</td>
<td style="text-align: right;">146.48</td>
<td style="text-align: right;">182.1</td>
<td style="text-align: left;">▁▇▅▅▃</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_4</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">119.15</td>
<td style="text-align: right;">29.24</td>
<td style="text-align: right;">67.0</td>
<td style="text-align: right;">96.05</td>
<td style="text-align: right;">110.2</td>
<td style="text-align: right;">143.40</td>
<td style="text-align: right;">182.6</td>
<td style="text-align: left;">▂▇▃▅▂</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Si tenim noms molt complicats, podem afegir abans<code>clean_names()</code> i posa tots els noms de columna en minuscula i sense espais, per fer més fàcil el codi, necessita paquet <code>janitor</code></p>
<p>Si us hi fixeu ara les variables que eren de categoria <code>character</code> per defecte, ara son <code>factor</code> i ja es veuen els nivells.</p>
<section id="filtrar-dades" class="level4">
<h4 class="anchored" data-anchor-id="filtrar-dades">Filtrar dades</h4>
<p>Si per exemple volem extreure només les dades de la generació “M” podriem crear un altre dataset de la següent manera:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>dams <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">==</span> <span class="st">"M"</span>)  </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(dams)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">dams</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">30</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">8</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">factor</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: factor</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 11%">
<col style="width: 16%">
<col style="width: 9%">
<col style="width: 10%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: left;">ordered</th>
<th style="text-align: right;">n_unique</th>
<th style="text-align: left;">top_counts</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">generation</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">M: 30, CA: 0, Cd1: 0, Cd2: 0</td>
</tr>
<tr class="even">
<td style="text-align: left;">group</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">G: 6, P: 6, PGS: 6, REF: 6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">sex</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">F: 30, M: 0</td>
</tr>
<tr class="even">
<td style="text-align: left;">rat</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">30</td>
<td style="text-align: left;">G1: 1, G3: 1, G4: 1, G6: 1</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 11%">
<col style="width: 16%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 6%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gene_1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0.90</td>
<td style="text-align: right;">110.77</td>
<td style="text-align: right;">28.24</td>
<td style="text-align: right;">61.3</td>
<td style="text-align: right;">94.15</td>
<td style="text-align: right;">105.0</td>
<td style="text-align: right;">130.55</td>
<td style="text-align: right;">161.7</td>
<td style="text-align: left;">▃▇▂▅▅</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_2</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0.90</td>
<td style="text-align: right;">115.63</td>
<td style="text-align: right;">32.90</td>
<td style="text-align: right;">67.8</td>
<td style="text-align: right;">89.60</td>
<td style="text-align: right;">105.2</td>
<td style="text-align: right;">137.10</td>
<td style="text-align: right;">175.5</td>
<td style="text-align: left;">▅▇▁▃▃</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gene_3</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0.93</td>
<td style="text-align: right;">118.07</td>
<td style="text-align: right;">30.69</td>
<td style="text-align: right;">50.7</td>
<td style="text-align: right;">94.45</td>
<td style="text-align: right;">110.8</td>
<td style="text-align: right;">145.80</td>
<td style="text-align: right;">173.7</td>
<td style="text-align: left;">▁▇▆▅▅</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0.90</td>
<td style="text-align: right;">116.40</td>
<td style="text-align: right;">28.53</td>
<td style="text-align: right;">71.2</td>
<td style="text-align: right;">96.40</td>
<td style="text-align: right;">108.8</td>
<td style="text-align: right;">134.15</td>
<td style="text-align: right;">177.1</td>
<td style="text-align: left;">▃▇▅▂▃</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>La funció <code>filter()</code> és molt interessant ja que podem fer-ho de una variable només un nivell amb <code>== "nom variable"</code> o de diferents nivells <code>%in%c("nom1", "nom2")</code> i alhora de diferents variables i diferents nivells</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>data <span class="sc">%&gt;%</span> </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation  <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"CA"</span>), group <span class="sc">==</span> <span class="st">"REF"</span> )  </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 16 × 8
   generation group sex   rat   gene_1 gene_2 gene_3 gene_4
   &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
 1 M          REF   F     R1     105     94     95.2  101  
 2 M          REF   F     R2      98.6  118.    98.1  110. 
 3 M          REF   F     R6     106.    99.9   88.9   NA  
 4 M          REF   F     R8     115.    89.4   88     96.7
 5 M          REF   F     R9      97.7  108.   108.    96.1
 6 M          REF   F     R10     97.7   87.8   NA     85.4
 7 CA         REF   F     R1.2    93.8  108.    94.4  101. 
 8 CA         REF   F     R1.3    94.4  105.   119.    93.2
 9 CA         REF   F     R1.6    93.6   NA     85.5  108. 
10 CA         REF   F     R2.6   112.   112.    78     NA  
11 CA         REF   F     R2.7   114.   114.   104.    95.5
12 CA         REF   M     R6.6    NA    118.    84.8  114. 
13 CA         REF   M     R6.7   103.   106.    NA     93.2
14 CA         REF   M     R8.2   125.    96.4  104.   118  
15 CA         REF   M     R8.3    93.6  106.   116.    99.6
16 CA         REF   M     R8.6    94.7  111.    99.3   85.7</code></pre>
</div>
</div>
<p>Fixeu-vos que en aquest cas m’ho ensenya directament. Això és degut a que no m’ho està guardant com a dataframe perquè no he posat <code>nom_dataset &lt;-</code> davant de les dades. Pot ser útil per comprovar si el teu codi fa el que vols abans de crear el dataframe.</p>
</section>
</section>
<section id="wide-format-vs-long-format" class="level3">
<h3 class="anchored" data-anchor-id="wide-format-vs-long-format">Wide format vs Long format</h3>
<p>Un concepte molt important a l’hora de utilitzar R és el wide dataset vs long dataset.</p>
<p><img src="images/Captura de pantalla 2026-01-16 162651.png" class="img-fluid" width="435"></p>
<p>En aquesta imatge tenim a l’<strong>esquerra</strong> el format <strong>wide</strong>, que és el que estem més acostumats a crear en excel, i a la <strong>dreta</strong> el format <strong>long</strong>, el qual és el que li va millor a R per la majoria de funcions. En el nostre cas li hem donat una taula en format wide ja que tenim una rata a cada fila, i tenim una columna per a cada variable de gen. Per a tenir-ho en format long volem una columna que ens digui de quin gen es tracta aquella observació i una altre columna amb el valor de la observació.</p>
<p>Es pot fer fàcilment amb la següent formula, i aprofitem per recordar les variables que son factors i indicar l’ordre en que volem que apareguin a posteriori:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>longdata <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="st">"gene_1"</span><span class="sc">:</span><span class="st">"gene_4"</span>, <span class="at">names_to=</span><span class="st">"gene"</span>, <span class="at">values_to =</span> <span class="st">"value"</span> ) <span class="sc">%&gt;%</span> </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> <span class="fu">factor</span>(group, </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>, <span class="st">"S"</span>, <span class="st">"PGS"</span>)), </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">generation =</span> <span class="fu">factor</span>(generation, </span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> <span class="fu">c</span>( <span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>, <span class="st">"CA"</span>)),</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">gene =</span> <span class="fu">factor</span>(gene,</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>))</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>                    )</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(longdata) <span class="co">#per visualitzar-ho</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  generation group sex   rat   gene   value
  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;  &lt;dbl&gt;
1 M          REF   F     R1    gene_1 105  
2 M          REF   F     R1    gene_2  94  
3 M          REF   F     R1    gene_3  95.2
4 M          REF   F     R1    gene_4 101  
5 M          REF   F     R2    gene_1  98.6
6 M          REF   F     R2    gene_2 118. </code></pre>
</div>
</div>
<ul>
<li><p><code>cols="gene_1":"gene_4"</code> ens indica quines són les columnes que contenen la variable a convertir</p></li>
<li><p><code>names_to="gene"</code> el nom que tindrà la nova columna que indiqui quina variable és</p></li>
<li><p><code>values_to = "value"</code> el nom que tindrà la columna amb els valors de la variable</p></li>
</ul>
<p>En el cas que a les dades els grups o el sexe els tinguis com a número enlloc de lletres, pots canviar-ho de la següent manera</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">sex =</span> <span class="fu">factor</span>(sex, </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"0"</span>, <span class="st">"1"</span>), <span class="co">#com està ara</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Male"</span>, <span class="st">"Female"</span>)) <span class="co"># a què correspon cada un</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Recomano guardar les dades en format long per si les necessites ja que si ho fessis a ma tindries feina…</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(writexl)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write_xlsx</span>(longdata,<span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"PCR_long.xlsx"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>De la mateixa manera que ara hem convertit de wide a long, en el cas de tenir un dataframe en format long i que el vulguem llegir de manera més fàcil, podem passar <strong>de long a wide.</strong></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>widedata <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span>  </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from=</span><span class="st">"gene"</span>, <span class="at">values_from=</span><span class="st">"value"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(widedata)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  generation group sex   rat   gene_1 gene_2 gene_3 gene_4
  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 M          REF   F     R1     105     94     95.2  101  
2 M          REF   F     R2      98.6  118.    98.1  110. 
3 M          REF   F     R6     106.    99.9   88.9   NA  
4 M          REF   F     R8     115.    89.4   88     96.7
5 M          REF   F     R9      97.7  108.   108.    96.1
6 M          REF   F     R10     97.7   87.8   NA     85.4</code></pre>
</div>
</div>
<ul>
<li><p><code>names_from="gene"</code> per a saber d’on treurà els noms de les noves columnes que crearà</p></li>
<li><p><code>values_from="value"</code> d’on treurà els valors de cada cel·la</p></li>
</ul>
<section id="problema-dels-na" class="level4">
<h4 class="anchored" data-anchor-id="problema-dels-na">Problema dels NA</h4>
<p>És molt important saber en quin format tenim les dades ja que R no treballa bé amb els valors que no existeixen (<code>NA</code>). Com us he comentat, jo ja he fet un pre-processat de les dades, per tant he eliminat els outliers i aquelles cel·les estan buides. Al fer <code>skim</code> veiem a <code>n_missing</code> si en les nostres variables tenim alguns <code>NA</code> .</p>
<p>Per a fer els càlculs, a R podem fer la funció <code>na.omit()</code> . El que fa és eliminar tota la fila en la que es trobi un <code>NA</code> , i per tant podrà fer els càlculs simplement sense aquella fila.</p>
<p>PROBLEMA: en format wide, al eliminar una observació, elimines totes les observacions de tots els gens. Per tant si no tens el valor d’un gen però si de tots els altres, els elimines tots i perds dades.</p>
<p>En canvi, si ho fas en format long, com que cada observació només és d’un gen i el seu valor, sí que et quedes amb la resta de gens, per tant perds menys informació.</p>
<p>Per això recomano sempre que es faci estadística fer-ho amb el format LONG de les dades.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(data) <span class="co">#dades originals</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> generation group    sex         rat          gene_1          gene_2     
 CA  :20    G  :26   F:85   G1.1   :  2   Min.   : 61.3   Min.   : 67.2  
 Cd1 :40    P  :26   M:55   G3.1   :  2   1st Qu.: 93.7   1st Qu.: 97.0  
 Cd21:50    PGS:36          G8.1   :  2   Median :106.7   Median :111.4  
 M   :30    REF:36          P2.1   :  2   Mean   :115.6   Mean   :119.7  
            S  :16          P3.1   :  2   3rd Qu.:139.2   3rd Qu.:141.2  
                            P4.1   :  2   Max.   :170.7   Max.   :192.4  
                            (Other):128   NA's   :13      NA's   :13     
     gene_3          gene_4      
 Min.   : 50.7   Min.   : 67.00  
 1st Qu.: 97.3   1st Qu.: 96.05  
 Median :109.0   Median :110.20  
 Mean   :119.5   Mean   :119.15  
 3rd Qu.:146.5   3rd Qu.:143.40  
 Max.   :182.1   Max.   :182.60  
 NA's   :10      NA's   :13      </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>longdata <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from=</span><span class="st">"gene"</span>, <span class="at">values_from=</span><span class="st">"value"</span>) <span class="sc">%&gt;%</span> <span class="co">#per comparar quantes elimina i per tant amb que fariem els anàlisis</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> generation group    sex         rat          gene_1          gene_2     
 M   :30    REF:36   F:85   G1.1   :  2   Min.   : 61.3   Min.   : 67.2  
 Cd1 :40    P  :26   M:55   G3.1   :  2   1st Qu.: 93.7   1st Qu.: 97.0  
 Cd21:50    G  :26          G8.1   :  2   Median :106.7   Median :111.4  
 CA  :20    S  :16          P2.1   :  2   Mean   :115.6   Mean   :119.7  
            PGS:36          P3.1   :  2   3rd Qu.:139.2   3rd Qu.:141.2  
                            P4.1   :  2   Max.   :170.7   Max.   :192.4  
                            (Other):128   NA's   :13      NA's   :13     
     gene_3          gene_4      
 Min.   : 50.7   Min.   : 67.00  
 1st Qu.: 97.3   1st Qu.: 96.05  
 Median :109.0   Median :110.20  
 Mean   :119.5   Mean   :119.15  
 3rd Qu.:146.5   3rd Qu.:143.40  
 Max.   :182.1   Max.   :182.60  
 NA's   :10      NA's   :13      </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>widedata <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code> generation group    sex         rat         gene_1          gene_2     
 M   :19    REF:28   F:58   G1.1   : 2   Min.   : 64.1   Min.   : 67.2  
 Cd1 :29    P  :17   M:39   G3.1   : 2   1st Qu.: 94.3   1st Qu.: 97.2  
 Cd21:36    G  :17          P4.1   : 2   Median :105.0   Median :109.8  
 CA  :13    S  :11          R1.1   : 2   Mean   :115.6   Mean   :119.9  
            PGS:24          R1.2   : 2   3rd Qu.:139.2   3rd Qu.:140.2  
                            R6.1   : 2   Max.   :170.7   Max.   :183.7  
                            (Other):85                                  
     gene_3          gene_4     
 Min.   : 50.7   Min.   : 67.0  
 1st Qu.: 97.9   1st Qu.: 96.1  
 Median :108.5   Median :106.2  
 Mean   :118.9   Mean   :119.1  
 3rd Qu.:146.4   3rd Qu.:144.3  
 Max.   :181.7   Max.   :182.6  
                                </code></pre>
</div>
</div>
<p>Com podeu veure al treballar amb les dades long els resultats són els mateixos que amb les dates originals i no hem eliminat cap rata sencera. Fixeu-vos també com les dades que hem aconseguit a partir de long estan ordenades com jo volia (li he indicat al crear el dataset) i no com R vol (com en el cas de les originals).</p>
</section>
<section id="exemple-random-accions-extra" class="level4">
<h4 class="anchored" data-anchor-id="exemple-random-accions-extra">Exemple random accions extra</h4>
<p>Quan estava re-aprenent-ho tot, em van fer crear un dataset nou amb moltes variables noves i altres canviades, i us ho poso d’exemple perquè sapigueu que aquestes coses es poden fer</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>cleanbeaches_new <span class="ot">&lt;-</span> cleanbeaches <span class="sc">%&gt;%</span>  </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(date, <span class="fu">c</span>(<span class="st">"day"</span>, <span class="st">"month"</span>, <span class="st">"year"</span>), <span class="at">remove=</span>F) <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">logbeachbugs =</span> <span class="fu">log</span>(beachbugs)) <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">beachbugsdiff =</span> beachbugs <span class="sc">-</span> <span class="fu">lag</span>(beachbugs)) <span class="sc">%&gt;%</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">buggier_all =</span> beachbugs <span class="sc">&gt;</span> <span class="fu">mean</span>(beachbugs, <span class="at">na.rm=</span>T)) <span class="sc">%&gt;%</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(site)<span class="sc">%&gt;%</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">buggier_site =</span> beachbugs <span class="sc">&gt;</span> <span class="fu">mean</span>(beachbugs, <span class="at">na.rm=</span>T))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li><p><code>cleanbeaches_new &lt;- cleanbeaches %&gt;%</code> crea un dataset nou a partir de les dades cleanbeaches i aplica les següents accions</p></li>
<li><p><code>separate(date, c("day", "month", "year"), remove=F) %&gt;%</code> tens una columna que es diu <code>date</code>, separa-la en tres columnes anomenades <code>"day", "month", "year"</code> , però no m’eliminis l’original</p></li>
<li><p><code>mutate(logbeachbugs = log(beachbugs)) %&gt;%</code> afegeix una columna nova que es digui <code>logbeachbugs</code> i contingui les dades de la columna que ja existeix i es diu <code>beachbugs</code> però aplica la funció <code>log(x)</code> als seus valors.</p></li>
<li><p><code>mutate(beachbugsdiff = beachbugs - lag(beachbugs)) %&gt;%</code> afegeix una columna nova fent la operació <code>x-lag(x)</code> de les dades de la columna beachbugs</p></li>
<li><p><code>mutate(buggier_all = beachbugs &gt; mean(beachbugs, na.rm=T)) %&gt;%</code> afegeix una columna on indiqui si aquella observació és major a la mitjana (i al fer la mitjana no tinguis en compte els NA: <code>na.rm=T</code>)</p></li>
<li><p><code>group_by(site)%&gt;% mutate(buggier_site = beachbugs &gt; mean(beachbugs, na.rm=T))</code> per començar el que facis ara agrupa-ho per site i llavors afegeix una columna que em digui si la observació és major a la mitjana només del site.</p>
<ul>
<li>Al haver fet <code>group_by()</code> el que es calculi a partir d’ara es calculara en els diferents sites, no a tot el dataset. Per evitar-ho despres s’ha de fer <code>ungroup()</code></li>
</ul></li>
</ul>
</section>
</section>
</section>
<section id="estadística-descriptiva" class="level2">
<h2 class="anchored" data-anchor-id="estadística-descriptiva">Estadística descriptiva</h2>
<p>La majoria de l’estadística descriptiva ja la mirem amb l’excel, però per comprovar que tot ho estem fent igual, podem demanar una taula amb la mitjana, la sd, l’error, la n i els límits.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>total_results <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span>   <span class="co">#longdata!,wide data també es podria, no recomano</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(generation, group, gene) <span class="sc">%&gt;%</span> <span class="co">#que ho faci tot dins de la mateixa                                               generació i mateix grup i per cada gen</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarise</span>(</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">across</span>(</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>           <span class="fu">where</span>(is.numeric),</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>          <span class="fu">list</span>(</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">n =</span> <span class="sc">~</span><span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(.x)),</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean =</span> <span class="sc">~</span><span class="fu">mean</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), </span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd =</span> <span class="sc">~</span><span class="fu">sd</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">error =</span> <span class="sc">~</span><span class="fu">sd</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">/</span> <span class="fu">sqrt</span>(<span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(.x))),</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">upper =</span> <span class="sc">~</span><span class="fu">mean</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> <span class="fu">sd</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">lower =</span> <span class="sc">~</span><span class="fu">mean</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">-</span> <span class="dv">2</span> <span class="sc">*</span> <span class="fu">sd</span>(.x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>          ),</span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>          <span class="at">.names =</span> <span class="st">"{.col}_{.fn}"</span></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>       )) <span class="sc">%&gt;%</span></span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>          <span class="fu">as.data.frame</span>() <span class="co">#ensenyar-ho com a dataframe</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'generation', 'group'. You can override
using the `.groups` argument.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(total_results)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(total_results)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  generation group   gene value_n value_mean  value_sd value_error value_upper
1          M   REF gene_1       6  103.45000  6.919754    2.824978    117.2895
2          M   REF gene_2       6   99.63333 11.881863    4.850750    123.3971
3          M   REF gene_3       5   95.66000  8.143893    3.642060    111.9478
4          M   REF gene_4       5   97.78000  8.799261    3.935149    115.3785
5          M     P gene_1       5  129.72000  9.279925    4.150108    148.2798
6          M     P gene_2       5  128.08000 15.767910    7.051624    159.6158
  value_lower
1    89.61049
2    75.86961
3    79.37221
4    80.18148
5   111.16015
6    96.54418</code></pre>
</div>
</div>
<p>Al tenir activat <code>na.rm=TRUE</code> s’eliminen els NA del que estàs calculant.</p>
</section>
<section id="inferència-estadística" class="level2">
<h2 class="anchored" data-anchor-id="inferència-estadística">Inferència estadística</h2>
<p>Per a saber millor el perquè i què fa cada test, recomano consultar el llibre <a href="https://learningstatisticswithr.com/book/" class="uri">https://learningstatisticswithr.com/book/</a>. Si ho voleu com algo rutinari, aquí teniu la taula del protocol que va fer la Mariona Camps al 2016 per al SPSS.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 16%">
<col style="width: 16%">
<col style="width: 16%">
<col style="width: 16%">
<col style="width: 16%">
<col style="width: 16%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Nº grups</th>
<th style="text-align: center;">Factors fixos (variables independents)</th>
<th style="text-align: center;">Segueix la normalitat?</th>
<th style="text-align: center;">Segueix la homogeneitat de variàncies?</th>
<th style="text-align: center;">Son dades aparellades?</th>
<th style="text-align: center;">Test</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Si/No</td>
<td style="text-align: center;">Si/No</td>
<td style="text-align: center;">T-test</td>
</tr>
<tr class="even">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Wilcoxon</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">U Mann Whitney</td>
</tr>
<tr class="even">
<td style="text-align: center;">1 (mesures repetides)</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Friedman</td>
</tr>
<tr class="odd">
<td style="text-align: center;">&gt;2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">Anova one-way (Posthoc -Bonferroni)</td>
</tr>
<tr class="even">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">No</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">Anova one-way (Posthoc -Dunnet T3 (SPSS); Games-Howell (R)</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">Kruskal Wallis (Posthoc Dunn -Bonferroni correction)</td>
</tr>
<tr class="even">
<td style="text-align: center;"></td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Anova two-way mesures repetides</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="even">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">Si</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">Anova two-way</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">No</td>
<td style="text-align: center;">Art ANOVA</td>
</tr>
</tbody>
</table>
<p>Sigui com sigui sempre hem de començar mirant la normalitat i homogenitat de variàncies de les dades.</p>
<p>Abans de començar a fer gràfics canviarem el “default theme” a classic.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_classic</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>També, al demanar-li View(), el separador decimal és en punt, però jo sempre treballo amb l’excel amb comes, per tant per poder-ho copiar directament:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>copy <span class="ot">&lt;-</span> results_test<span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">where</span>(is.numeric), <span class="sc">~</span> <span class="fu">ifelse</span>(</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">is.na</span>(.),</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">""</span>,</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">format</span>(., <span class="at">decimal.mark =</span> <span class="st">","</span>, <span class="at">scientific =</span> <span class="cn">FALSE</span>))))</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(copy)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="normalitat" class="level3">
<h3 class="anchored" data-anchor-id="normalitat">Normalitat</h3>
<p>En el cas que la el número de grups diferents sigui de n&lt;50 s’ha de fer <strong>Shapiro-Wilk</strong>. En el cas que la n&gt;50, Kolmogorov-Smirnov, el qual no explicaré ja que no l’hem de fer pràcticament mai.</p>
<p>Per a la normalitat primer es comprova amb un <strong>QQplot</strong>, de manera visual</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>QQplot <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span> </span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, group, gene) <span class="sc">%&gt;%</span> <span class="co">#com separaras a l'anàlisi posterior</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">sample =</span> value)) <span class="sc">+</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_qq</span>() <span class="sc">+</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_qq_line</span>() <span class="sc">+</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(generation <span class="sc">+</span> group <span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"QQ plots of value by generation, group and gene"</span>,</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Sample Quantiles"</span>,</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Theoretical Quantiles"</span>)</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>QQplot</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Tutorial-Rstudio_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>En el cas de voler només una generació canviar el <code>group_by(generation)</code> per <code>filter(generation == "CA")</code>.</p>
<p>Com que no som experts en estadística, nosaltres també farem el test de Shapiro-Wilk per a que ens digui si són normals (p&gt;0.05) o no són normals (p&lt;0.05)</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>shapiro_by_group <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, group, gene) <span class="sc">%&gt;%</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># SPSS-style df = size of the sample used</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">df =</span> <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(value)),</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Shapiro–Wilk test (only runs if df ≥ 3, same as SPSS)</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">W =</span> <span class="cf">if</span>(df <span class="sc">&gt;=</span> <span class="dv">3</span>) <span class="fu">shapiro.test</span>(value)<span class="sc">$</span>statistic <span class="cf">else</span> <span class="cn">NA_real_</span>,</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">p_value =</span> <span class="cf">if</span>(df <span class="sc">&gt;=</span> <span class="dv">3</span>) <span class="fu">shapiro.test</span>(value)<span class="sc">$</span>p.value <span class="cf">else</span> <span class="cn">NA_real_</span>,</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">.groups =</span> <span class="st">"drop"</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>  )<span class="sc">%&gt;%</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(                  <span class="co">#per afegir columna de * significació, es pot treure</span></span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">signif =</span> <span class="fu">case_when</span>(</span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>      <span class="fu">is.na</span>(p_value)        <span class="sc">~</span> <span class="st">"ns"</span>,</span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>      p_value <span class="sc">&lt;</span> <span class="fl">0.001</span>       <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>      p_value <span class="sc">&lt;</span> <span class="fl">0.01</span>        <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a>      p_value <span class="sc">&lt;</span> <span class="fl">0.05</span>        <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a>      p_value <span class="sc">&lt;</span> <span class="fl">0.1</span>  <span class="sc">~</span> <span class="st">"·"</span>,</span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span>                  <span class="sc">~</span> <span class="st">"ns"</span></span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a>    )) <span class="co"># A R la significació es veu més neta</span></span>
<span id="cb44-22"><a href="#cb44-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-23"><a href="#cb44-23" aria-hidden="true" tabindex="-1"></a>shapiro_by_group</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 64 × 7
   generation group gene      df     W p_value signif
   &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt;  &lt;int&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt; 
 1 M          REF   gene_1     6 0.853  0.165  ns    
 2 M          REF   gene_2     6 0.924  0.534  ns    
 3 M          REF   gene_3     5 0.915  0.498  ns    
 4 M          REF   gene_4     5 0.970  0.876  ns    
 5 M          P     gene_1     5 0.738  0.0228 *     
 6 M          P     gene_2     5 0.735  0.0215 *     
 7 M          P     gene_3     6 0.978  0.941  ns    
 8 M          P     gene_4     6 0.856  0.175  ns    
 9 M          G     gene_1     5 0.858  0.222  ns    
10 M          G     gene_2     5 0.970  0.876  ns    
# ℹ 54 more rows</code></pre>
</div>
</div>
<ul>
<li>En aquest cas el gen_1 i gen_2 en alguns grups i algunes generacions no segueixen la normalitat, llavors en el nostre cas decidiriem que aplicariem tests de no normal en tots els casos, però es podria discutir a veure quin aplicar.</li>
</ul>
</section>
<section id="homogeneïtat-de-les-variàncies" class="level3">
<h3 class="anchored" data-anchor-id="homogeneïtat-de-les-variàncies">Homogeneïtat de les variàncies</h3>
<p>Test <strong>Levene</strong> per a que ens digui si les variàncies són homogènies (p&gt;0.05) o no són homogènies (p&lt;0.05). Es fa per tots els grups en general, no cal separar grups. En el cas de SPSS es feia amb l’Anova, a R el codi seria:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>levene_long <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, gene) <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_modify</span>(<span class="sc">~</span> {</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># eliminem NA</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    d <span class="ot">&lt;-</span> .x[<span class="sc">!</span><span class="fu">is.na</span>(.x<span class="sc">$</span>value) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(.x<span class="sc">$</span>group), ]</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># --- Levene clàssics ---</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>    lev_mean   <span class="ot">&lt;-</span> <span class="fu">leveneTest</span>(value <span class="sc">~</span> group, d, <span class="at">center =</span> mean)</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>    lev_median <span class="ot">&lt;-</span> <span class="fu">leveneTest</span>(value <span class="sc">~</span> group, d, <span class="at">center =</span> median)</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">method =</span> <span class="fu">c</span>(<span class="st">"Based on Mean"</span>,<span class="st">"Based on Median"</span>),</span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">F =</span> <span class="fu">c</span>(lev_mean[<span class="dv">1</span>, <span class="st">"F value"</span>],lev_median[<span class="dv">1</span>, <span class="st">"F value"</span>]),</span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">df1 =</span> <span class="fu">c</span>(lev_mean[<span class="dv">1</span>, <span class="st">"Df"</span>],lev_median[<span class="dv">1</span>, <span class="st">"Df"</span>]     ),</span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">df2 =</span> <span class="fu">c</span>(lev_mean[<span class="dv">2</span>, <span class="st">"Df"</span>],lev_median[<span class="dv">2</span>, <span class="st">"Df"</span>]),</span>
<span id="cb46-18"><a href="#cb46-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">p.value =</span> <span class="fu">c</span>(lev_mean[<span class="dv">1</span>, <span class="st">"Pr(&gt;F)"</span>],lev_median[<span class="dv">1</span>, <span class="st">"Pr(&gt;F)"</span>]))</span>
<span id="cb46-19"><a href="#cb46-19" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb46-20"><a href="#cb46-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb46-21"><a href="#cb46-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(                  <span class="co">#per afegir columna de * significació, es pot treure</span></span>
<span id="cb46-22"><a href="#cb46-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">signif =</span> <span class="fu">case_when</span>(</span>
<span id="cb46-23"><a href="#cb46-23" aria-hidden="true" tabindex="-1"></a>      <span class="fu">is.na</span>(p.value)        <span class="sc">~</span> <span class="st">"ns"</span>,</span>
<span id="cb46-24"><a href="#cb46-24" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span>       <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb46-25"><a href="#cb46-25" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span>        <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb46-26"><a href="#cb46-26" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span>        <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb46-27"><a href="#cb46-27" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span>  <span class="sc">~</span> <span class="st">"·"</span>,</span>
<span id="cb46-28"><a href="#cb46-28" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span>                  <span class="sc">~</span> <span class="st">"ns"</span></span>
<span id="cb46-29"><a href="#cb46-29" aria-hidden="true" tabindex="-1"></a>    )) <span class="co"># A R la significació es veu més neta</span></span>
<span id="cb46-30"><a href="#cb46-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-31"><a href="#cb46-31" aria-hidden="true" tabindex="-1"></a>levene_long</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 32 × 8
   generation gene   method               F   df1   df2 p.value signif
   &lt;fct&gt;      &lt;fct&gt;  &lt;chr&gt;            &lt;dbl&gt; &lt;int&gt; &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt; 
 1 M          gene_1 Based on Mean   2.54       4    22  0.0686 ·     
 2 M          gene_1 Based on Median 1.52       4    22  0.232  ns    
 3 M          gene_2 Based on Mean   0.0455     4    22  0.996  ns    
 4 M          gene_2 Based on Median 0.0161     4    22  0.999  ns    
 5 M          gene_3 Based on Mean   0.638      4    23  0.641  ns    
 6 M          gene_3 Based on Median 0.576      4    23  0.683  ns    
 7 M          gene_4 Based on Mean   0.395      4    22  0.810  ns    
 8 M          gene_4 Based on Median 0.330      4    22  0.855  ns    
 9 Cd1        gene_1 Based on Mean   0.235      3    33  0.871  ns    
10 Cd1        gene_1 Based on Median 0.207      3    33  0.891  ns    
# ℹ 22 more rows</code></pre>
</div>
</div>
<ul>
<li>En aquest cas el gen_1 i gen_2 en alguns grups i algunes generacions les variàncies no son homogenies, llavors en el nostre cas decidiriem que aplicariem tests de no homogènies en tots els casos, però es podria discutir a veure quin aplicar.</li>
</ul>
<p>Aquí també tenim una diferència respecte a SPSS, ja que levene no permet NA, llavors en el format wide elimina tota la observació. Al treballar en long només eliminem la observació del gen concret per tant mantenim més dades per als altres gens. Si es vol comprovar o fer, previ al test LEVENE fer:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>longdata2 <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="st">"gene_1"</span><span class="sc">:</span><span class="st">"gene_4"</span>, <span class="at">names_to=</span><span class="st">"gene"</span>, <span class="at">values_to =</span> <span class="st">"value"</span> ) <span class="sc">%&gt;%</span> </span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> <span class="fu">factor</span>(group, </span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>, <span class="st">"S"</span>, <span class="st">"PGS"</span>)), </span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">generation =</span> <span class="fu">factor</span>(generation, </span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> <span class="fu">c</span>( <span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>, <span class="st">"CA"</span>)),</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">gene =</span> <span class="fu">factor</span>(gene,</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>                    <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>)))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>A part SPSS ho feia centrat en la mitjana, mediana (amb R ho fem) i “trimmed mean” i “median and with adjusted df”. Aquestes últimes no les calculem, però no ho crec necessari ja que normalment els resultats son o tots no significatius o tots significatius.</p>
</section>
<section id="independent-samples-t-test" class="level3">
<h3 class="anchored" data-anchor-id="independent-samples-t-test">Independent samples T-test</h3>
<p>Per a comparacions de mitjanes on hi ha dos grups i les mostres son normals: <strong>T-test de mostres independents</strong></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>results_ttest <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">==</span> <span class="st">"CA"</span>) <span class="sc">%&gt;%</span> <span class="co">#només aquesta té dos grups</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, gene) <span class="sc">%&gt;%</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># fer t-test i guardar el resultat</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">t_res =</span> <span class="fu">list</span>(<span class="fu">t.test</span>(value <span class="sc">~</span> group, <span class="at">var.equal =</span> T, <span class="at">alternative =</span> <span class="st">"two.sided"</span>, <span class="at">conf.level=</span><span class="fl">0.95</span>)), <span class="co">#canviar a F si equal variances not assumed ; paired = F, es separen les variables amb ,</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">.groups =</span> <span class="st">"drop"</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">t =</span> t_res<span class="sc">$</span>statistic,</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">df =</span> t_res<span class="sc">$</span>parameter,</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_diff =</span> <span class="fu">diff</span>(t_res<span class="sc">$</span>estimate),  <span class="co"># Mean difference (group1 - group2)</span></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">std_error =</span> <span class="fu">abs</span>(mean_diff) <span class="sc">/</span> <span class="fu">abs</span>(t),  <span class="co"># Std. Error Difference</span></span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">p_one_sided =</span> t_res<span class="sc">$</span>p.value <span class="sc">/</span> <span class="dv">2</span>,</span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">p_two_sided =</span> t_res<span class="sc">$</span>p.value,</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">sig =</span> <span class="fu">case_when</span>(</span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">is.na</span>(t_res<span class="sc">$</span>p.value)        <span class="sc">~</span> <span class="st">"ns"</span>,</span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>      t_res<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.001</span>       <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a>      t_res<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.01</span>        <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a>      t_res<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.05</span>        <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a>      t_res<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.1</span>  <span class="sc">~</span> <span class="st">"·"</span>,</span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span>                  <span class="sc">~</span> <span class="st">"ns"</span>), <span class="co">#no cal, per visualitzar-ho millor</span></span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">conf_low =</span> t_res<span class="sc">$</span>conf.int[<span class="dv">1</span>],</span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">conf_high =</span> t_res<span class="sc">$</span>conf.int[<span class="dv">2</span>]</span>
<span id="cb49-26"><a href="#cb49-26" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb49-27"><a href="#cb49-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(generation, gene, t, df, p_one_sided, p_two_sided, sig, mean_diff, std_error, conf_low, conf_high)</span>
<span id="cb49-28"><a href="#cb49-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-29"><a href="#cb49-29" aria-hidden="true" tabindex="-1"></a>results_ttest</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 11
# Rowwise: 
  generation gene        t    df   p_one_sided  p_two_sided sig   mean_diff
  &lt;fct&gt;      &lt;fct&gt;   &lt;dbl&gt; &lt;dbl&gt;         &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;
1 CA         gene_1 -10.7     16 0.00000000566 0.0000000113 ***        53.6
2 CA         gene_2  -9.83    15 0.0000000312  0.0000000624 ***        58.3
3 CA         gene_3  -9.17    17 0.0000000270  0.0000000540 ***        57.4
4 CA         gene_4 -10.8     15 0.00000000906 0.0000000181 ***        59.9
# ℹ 3 more variables: std_error &lt;dbl&gt;, conf_low &lt;dbl&gt;, conf_high &lt;dbl&gt;</code></pre>
</div>
</div>
<ul>
<li><p>En el cas que la homogeneïtat de variàncies no es compleixi, modificar <code>var.equal = F</code>.</p></li>
<li><p>En el cas de dades aparellades afegir <code>paired = T</code> i eliminar <code>var.equal</code></p></li>
<li><p>En el cas que tenir vectors canvia la formula <code>t.test(value, group)</code>. No acostuma a ser el nostre cas, vigilar!!</p></li>
</ul>
</section>
<section id="one-way-anova" class="level3">
<h3 class="anchored" data-anchor-id="one-way-anova">One-way ANOVA</h3>
<p>Quan tenim més de dos grups i les dades són normals i homogènies fem el <strong>One-way ANOVA</strong>, amb el post-hoc ajustat per <strong>Bonferroni</strong>.</p>
<p>Aquest codi només mostra el resultat</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>longdata <span class="sc">%&gt;%</span> </span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_name =</span> gene) <span class="sc">%&gt;%</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_walk</span>(<span class="sc">~</span> {</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">====================</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"GENE:"</span>, <span class="fu">unique</span>(.x<span class="sc">$</span>gene_name), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Generation:"</span>, <span class="fu">unique</span>(.x<span class="sc">$</span>generation), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">summary</span>(<span class="fu">aov</span>(value <span class="sc">~</span> group, <span class="at">data =</span> .x)))</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>  })</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
====================
GENE: 1 
Generation: 1 2 3 
             Df Sum Sq Mean Sq F value Pr(&gt;F)    
group         4  67709   16927   120.8 &lt;2e-16 ***
Residuals   104  14572     140                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
11 observations deleted due to missingness

====================
GENE: 2 
Generation: 1 2 3 
             Df Sum Sq Mean Sq F value Pr(&gt;F)    
group         4  74249   18562   93.29 &lt;2e-16 ***
Residuals   105  20892     199                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
10 observations deleted due to missingness

====================
GENE: 3 
Generation: 1 2 3 
             Df Sum Sq Mean Sq F value Pr(&gt;F)    
group         4  75457   18864   106.1 &lt;2e-16 ***
Residuals   106  18840     178                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
9 observations deleted due to missingness

====================
GENE: 4 
Generation: 1 2 3 
             Df Sum Sq Mean Sq F value Pr(&gt;F)    
group         4  69539   17385   95.81 &lt;2e-16 ***
Residuals   105  19052     181                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
10 observations deleted due to missingness</code></pre>
</div>
</div>
<p>Aquest altre crea una taula per copiar a l’excel</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstatix)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>results_anova <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>)) <span class="sc">%&gt;%</span> <span class="co">#seleccionar aquelles que sigui                                                         el test correcte</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, gene) <span class="sc">%&gt;%</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_map</span>(<span class="sc">~</span> {</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>    model <span class="ot">&lt;-</span> <span class="fu">aov</span>(value <span class="sc">~</span> group, <span class="at">data =</span> .x)</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># taula tidy del model</span></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>    tidy_model <span class="ot">&lt;-</span> <span class="fu">tidy</span>(model)<span class="sc">%&gt;%</span> </span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">gene =</span> .y<span class="sc">$</span>gene) <span class="sc">%&gt;%</span> </span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">generation =</span> .y<span class="sc">$</span>generation)</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># fila TOTAL</span></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>    total_row <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">term =</span> <span class="st">"Total"</span>,</span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">df =</span> <span class="fu">sum</span>(tidy_model<span class="sc">$</span>df, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">sumsq =</span> <span class="fu">sum</span>(tidy_model<span class="sc">$</span>sumsq, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">meansq =</span> <span class="cn">NA_real_</span>,</span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">statistic =</span> <span class="cn">NA_real_</span>,</span>
<span id="cb53-19"><a href="#cb53-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">p.value =</span> <span class="cn">NA_real_</span>,</span>
<span id="cb53-20"><a href="#cb53-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">gene =</span> .y<span class="sc">$</span>gene</span>
<span id="cb53-21"><a href="#cb53-21" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb53-22"><a href="#cb53-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">bind_rows</span>(tidy_model, total_row) </span>
<span id="cb53-23"><a href="#cb53-23" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb53-24"><a href="#cb53-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span></span>
<span id="cb53-25"><a href="#cb53-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(generation, gene, <span class="fu">everything</span>())   <span class="sc">%&gt;%</span></span>
<span id="cb53-26"><a href="#cb53-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb53-27"><a href="#cb53-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">signif =</span> <span class="fu">case_when</span>(</span>
<span id="cb53-28"><a href="#cb53-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">is.na</span>(p.value)  <span class="sc">~</span> <span class="st">"ns"</span>,</span>
<span id="cb53-29"><a href="#cb53-29" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb53-30"><a href="#cb53-30" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span>  <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb53-31"><a href="#cb53-31" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span>  <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb53-32"><a href="#cb53-32" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span>  <span class="sc">~</span> <span class="st">"·"</span>,</span>
<span id="cb53-33"><a href="#cb53-33" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span>            <span class="sc">~</span> <span class="st">"ns"</span>))</span>
<span id="cb53-34"><a href="#cb53-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-35"><a href="#cb53-35" aria-hidden="true" tabindex="-1"></a>results_anova</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 36 × 9
   generation gene   term         df  sumsq meansq statistic  p.value signif
   &lt;fct&gt;      &lt;fct&gt;  &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt; 
 1 M          gene_1 group         4 17584.  4396.      30.8  1.01e-8 ***   
 2 M          gene_1 Residuals    22  3144.   143.      NA   NA       ns    
 3 &lt;NA&gt;       gene_1 Total        26 20728.    NA       NA   NA       ns    
 4 M          gene_2 group         4 24297.  6074.      34.8  3.23e-9 ***   
 5 M          gene_2 Residuals    22  3843.   175.      NA   NA       ns    
 6 &lt;NA&gt;       gene_2 Total        26 28140.    NA       NA   NA       ns    
 7 M          gene_3 group         4 21158.  5290.      28.5  1.29e-8 ***   
 8 M          gene_3 Residuals    23  4266.   185.      NA   NA       ns    
 9 &lt;NA&gt;       gene_3 Total        27 25425.    NA       NA   NA       ns    
10 M          gene_4 group         4 16595.  4149.      20.0  4.52e-7 ***   
# ℹ 26 more rows</code></pre>
</div>
</div>
<p>Si fem cada gen per separat el codi és molt més senzill, però tenim més feina manual</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>dades_anova <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span> </span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">==</span> <span class="st">"M"</span>, gene <span class="sc">==</span> <span class="st">"gene_1"</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>anova <span class="ot">&lt;-</span> <span class="fu">aov</span>(value <span class="sc">~</span> group,dades_anova)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(anova)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "aov" "lm" </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(anova)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Df Sum Sq Mean Sq F value   Pr(&gt;F)    
group        4  17584    4396   30.76 1.01e-08 ***
Residuals   22   3144     143                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
3 observations deleted due to missingness</code></pre>
</div>
</div>
<section id="post-hoc-ajustat-per-bonferroni" class="level4">
<h4 class="anchored" data-anchor-id="post-hoc-ajustat-per-bonferroni">Post-hoc ajustat per Bonferroni</h4>
<p>En el cas que doni diferències significatives, passem a fer el post-hoc.</p>
<ul>
<li>Al fer emmeans: Aquest codi agafa cada model ANOVA, calcula les mitjanes ajustades, fa totes les comparacions parell a parell amb Bonferroni, i retorna estimacions + intervals de confiança + p-valors.</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(emmeans)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to emmeans.
Caution: You lose important information if you filter this package's results.
See '? untidy'</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>posthoc_results <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span> </span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, gene) <span class="sc">%&gt;%</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nest</span>() <span class="sc">%&gt;%</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">aov_model =</span> <span class="fu">map</span>(data, <span class="sc">~</span> <span class="fu">aov</span>(value <span class="sc">~</span> group, <span class="at">data =</span> .x)),</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">posthoc =</span> <span class="fu">map</span>(</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>      aov_model,</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>      <span class="sc">~</span> <span class="fu">emmeans</span>(.x, <span class="sc">~</span> group) <span class="sc">%&gt;%</span></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">contrast</span>(<span class="st">"pairwise"</span>, <span class="at">adjust =</span> <span class="st">"bonferroni"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">summary</span>(<span class="at">level =</span> <span class="fl">0.95</span>,<span class="at">infer =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span>   <span class="co"># IC + p-value</span></span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>())) <span class="sc">%&gt;%</span></span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(generation, gene, posthoc) <span class="sc">%&gt;%</span></span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(posthoc) <span class="sc">%&gt;%</span></span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">signif =</span> <span class="fu">case_when</span>(</span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a>      <span class="fu">is.na</span>(p.value)  <span class="sc">~</span> <span class="st">"ns"</span>,</span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span>  <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span>  <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span>  <span class="sc">~</span> <span class="st">"·"</span>,</span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span>            <span class="sc">~</span> <span class="st">"ns"</span>))   </span>
<span id="cb61-23"><a href="#cb61-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-24"><a href="#cb61-24" aria-hidden="true" tabindex="-1"></a>posthoc_results</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 104 × 11
# Groups:   generation, gene [12]
   generation gene   contrast  estimate    SE    df lower.CL upper.CL t.ratio
   &lt;fct&gt;      &lt;fct&gt;  &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;
 1 M          gene_1 REF - P     -26.3   7.24    22   -48.8    -3.69    -3.63
 2 M          gene_1 REF - G      24.4   7.24    22     1.81   47.0      3.37
 3 M          gene_1 REF - S       9.55  6.90    22   -12.0    31.1      1.38
 4 M          gene_1 REF - PGS   -49.1   7.24    22   -71.7   -26.5     -6.78
 5 M          gene_1 P - G        50.7   7.56    22    27.1    74.2      6.70
 6 M          gene_1 P - S        35.8   7.24    22    13.2    58.4      4.95
 7 M          gene_1 P - PGS     -22.8   7.56    22   -46.4     0.741   -3.02
 8 M          gene_1 G - S       -14.8   7.24    22   -37.4     7.74    -2.05
 9 M          gene_1 G - PGS     -73.5   7.56    22   -97.1   -49.9     -9.72
10 M          gene_1 S - PGS     -58.7   7.24    22   -81.2   -36.1     -8.10
# ℹ 94 more rows
# ℹ 2 more variables: p.value &lt;dbl&gt;, signif &lt;chr&gt;</code></pre>
</div>
</div>
<p>Si ho fèssim per cada gen alhora ho podriem fer amb aquestes fòrmules:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pairwise.t.test</span>( <span class="at">x =</span> dades_anova<span class="sc">$</span>value,   <span class="co"># outcome variable</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">g =</span> dades_anova<span class="sc">$</span>group_name,        <span class="co"># grouping variable</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">p.adjust.method =</span> <span class="st">"none"</span>    <span class="co"># which correction to use?</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="co">#o</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lsr)</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a><span class="fu">posthocPairwiseT</span>( <span class="at">x =</span> results_anova, <span class="at">p.adjust.method =</span> <span class="st">"bonferroni"</span> )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="post-hoc-games-howell" class="level4">
<h4 class="anchored" data-anchor-id="post-hoc-games-howell">Post-hoc Games-Howell</h4>
<p>En el cas de dades normals però no homogènies, amb l’SPSS faríem Dunnet T3. En R no existeix així que es fa el Games-Howell.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstatix)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>posthoc_gameshowell <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">==</span> <span class="st">"M"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">games_howell_test</span>(value <span class="sc">~</span> group)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>posthoc_gameshowell</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40 × 9
   gene   .y.   group1 group2 estimate conf.low conf.high     p.adj p.adj.signif
 * &lt;fct&gt;  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;       
 1 gene_1 value REF    P         26.3      8.52     44.0  0.007     **          
 2 gene_1 value REF    G        -24.4    -54.3       5.55 0.105     ns          
 3 gene_1 value REF    S         -9.55   -36.1      17.0  0.697     ns          
 4 gene_1 value REF    PGS       49.1     34.9      63.4  0.0000101 ****        
 5 gene_1 value P      G        -50.7    -80.7     -20.6  0.004     **          
 6 gene_1 value P      S        -35.8    -63.1      -8.50 0.012     *           
 7 gene_1 value P      PGS       22.8      4.58     41.1  0.017     *           
 8 gene_1 value G      S         14.8    -18.2      47.9  0.578     ns          
 9 gene_1 value G      PGS       73.5     43.6     103.   0.000744  ***         
10 gene_1 value S      PGS       58.7     32.0      85.3  0.00066   ***         
# ℹ 30 more rows</code></pre>
</div>
</div>
</section>
</section>
<section id="u-mann-whitney" class="level3">
<h3 class="anchored" data-anchor-id="u-mann-whitney">U-Mann Whitney</h3>
<p>Quan tenim dos grups i les dades no són ni normals ni homogènies fem el test de <strong>U-Mann Whitney</strong>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>results_umw <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">==</span> <span class="st">"CA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wilcox_test</span>(value <span class="sc">~</span> group) <span class="co">#el que dona R per defecte</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>results_umw</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 8
  gene   .y.   group1 group2    n1    n2 statistic         p
* &lt;fct&gt;  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 gene_1 value REF    PGS        9     9         0 0.000409 
2 gene_2 value REF    PGS        9     8         0 0.0000823
3 gene_3 value REF    PGS        9    10         0 0.0000217
4 gene_4 value REF    PGS        9     8         0 0.000631 </code></pre>
</div>
</div>
</section>
<section id="kruskall-wallis" class="level3">
<h3 class="anchored" data-anchor-id="kruskall-wallis">Kruskall-Wallis</h3>
<p>Quan tenim més de dos grups i les dades no són ni normals ni homogènies fem el test de <strong>Kruskall-Wallis</strong>.</p>
<p>Només l’imprimeix</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>longdata <span class="sc">%&gt;%</span> </span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">==</span> <span class="st">"M"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_name =</span> gene) <span class="sc">%&gt;%</span> </span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_walk</span>(<span class="sc">~</span> {</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">====================</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"GENE:"</span>, <span class="fu">unique</span>(.x<span class="sc">$</span>gene_name), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">kruskal.test</span>(value <span class="sc">~</span> group, <span class="at">data =</span> .x))</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>  })</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
====================
GENE: 1 

    Kruskal-Wallis rank sum test

data:  value by group
Kruskal-Wallis chi-squared = 22.311, df = 4, p-value = 0.0001738


====================
GENE: 2 

    Kruskal-Wallis rank sum test

data:  value by group
Kruskal-Wallis chi-squared = 19.741, df = 4, p-value = 0.0005618


====================
GENE: 3 

    Kruskal-Wallis rank sum test

data:  value by group
Kruskal-Wallis chi-squared = 21.543, df = 4, p-value = 0.0002471


====================
GENE: 4 

    Kruskal-Wallis rank sum test

data:  value by group
Kruskal-Wallis chi-squared = 18.413, df = 4, p-value = 0.001025</code></pre>
</div>
</div>
<p>Per a copiar-lo a l’excel.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>results_kruskal <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, gene) <span class="sc">%&gt;%</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_map</span>(<span class="sc">~</span> {</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Kruskal–Wallis test</span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>    model <span class="ot">&lt;-</span> <span class="fu">kruskal.test</span>(value <span class="sc">~</span> group, <span class="at">data =</span> .x)</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># tidy del test</span></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>    tidy_model <span class="ot">&lt;-</span> <span class="fu">tidy</span>(model) <span class="sc">%&gt;%</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(</span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">generation =</span> .y<span class="sc">$</span>generation,</span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">gene =</span> .y<span class="sc">$</span>gene,</span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">df =</span> parameter </span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">%&gt;%</span></span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">select</span>(generation, gene,df,statistic,p.value)</span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">signif =</span> <span class="fu">case_when</span>(</span>
<span id="cb70-19"><a href="#cb70-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">is.na</span>(p.value)  <span class="sc">~</span> <span class="st">"ns"</span>,</span>
<span id="cb70-20"><a href="#cb70-20" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">~</span> <span class="st">"***"</span>,</span>
<span id="cb70-21"><a href="#cb70-21" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.01</span>  <span class="sc">~</span> <span class="st">"**"</span>,</span>
<span id="cb70-22"><a href="#cb70-22" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.05</span>  <span class="sc">~</span> <span class="st">"*"</span>,</span>
<span id="cb70-23"><a href="#cb70-23" aria-hidden="true" tabindex="-1"></a>      p.value <span class="sc">&lt;</span> <span class="fl">0.1</span>  <span class="sc">~</span> <span class="st">"·"</span>,</span>
<span id="cb70-24"><a href="#cb70-24" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span>            <span class="sc">~</span> <span class="st">"ns"</span>))</span>
<span id="cb70-25"><a href="#cb70-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-26"><a href="#cb70-26" aria-hidden="true" tabindex="-1"></a>results_kruskal</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 6
   generation gene      df statistic     p.value signif
   &lt;fct&gt;      &lt;fct&gt;  &lt;int&gt;     &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; 
 1 M          gene_1     4      22.3 0.000174    ***   
 2 M          gene_2     4      19.7 0.000562    ***   
 3 M          gene_3     4      21.5 0.000247    ***   
 4 M          gene_4     4      18.4 0.00102     **    
 5 Cd1        gene_1     3      29.9 0.00000145  ***   
 6 Cd1        gene_2     3      26.0 0.00000967  ***   
 7 Cd1        gene_3     3      30.0 0.00000136  ***   
 8 Cd1        gene_4     3      28.8 0.00000242  ***   
 9 Cd21       gene_1     4      31.2 0.00000284  ***   
10 Cd21       gene_2     4      32.6 0.00000143  ***   
11 Cd21       gene_3     4      27.1 0.0000186   ***   
12 Cd21       gene_4     4      35.2 0.000000430 ***   </code></pre>
</div>
</div>
<p>Si son significatius, fem el post-hoc Dunn.</p>
<section id="post-hoc-dunn-amb-correcció-de-bonferroni" class="level4">
<h4 class="anchored" data-anchor-id="post-hoc-dunn-amb-correcció-de-bonferroni">Post-hoc Dunn amb correcció de Bonferroni</h4>
<p>Quan les dades no siguin ni normals ni homogènies, es fa el test post-hoc de Dunn amb la correcció de Bonferroni.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>posthoc_dunn <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation <span class="sc">==</span> <span class="st">"M"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_name =</span> gene) <span class="sc">%&gt;%</span></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene_name) <span class="sc">%&gt;%</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dunn_test</span>(</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>    value <span class="sc">~</span> group,</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">p.adjust.method =</span> <span class="st">"bonferroni"</span>)</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>posthoc_dunn</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40 × 10
   gene_name .y.   group1 group2    n1    n2 statistic         p    p.adj
 * &lt;fct&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;int&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
 1 gene_1    value REF    P          6     5     1.46  0.145     1       
 2 gene_1    value REF    G          6     5    -1.75  0.0803    0.803   
 3 gene_1    value REF    S          6     6    -0.946 0.344     1       
 4 gene_1    value REF    PGS        6     5     2.50  0.0125    0.125   
 5 gene_1    value P      G          5     5    -3.07  0.00214   0.0214  
 6 gene_1    value P      S          5     6    -2.36  0.0183    0.183   
 7 gene_1    value P      PGS        5     5     0.997 0.319     1       
 8 gene_1    value G      S          5     6     0.847 0.397     1       
 9 gene_1    value G      PGS        5     5     4.07  0.0000476 0.000476
10 gene_1    value S      PGS        6     5     3.40  0.000672  0.00672 
# ℹ 30 more rows
# ℹ 1 more variable: p.adj.signif &lt;chr&gt;</code></pre>
</div>
</div>
</section>
</section>
<section id="art-anova" class="level3">
<h3 class="anchored" data-anchor-id="art-anova">Art ANOVA</h3>
<p>Quan tenim més de dos grups, les dades no són ni normals ni homogènies i hi ha dues variables independents, és fa <strong>Art ANOVA</strong>. És molt poc comú necessitar aquest test.</p>
<p>Aquest test l’ha fet amb R la Sofía, com que no s’acostuma a necessitar no l’he generalitzat, però us adjunto l’script per si us interessa.</p>
<p>2 factors</p>
<p>3 factors</p>
</section>
</section>
<section id="correlacions" class="level2">
<h2 class="anchored" data-anchor-id="correlacions">Correlacions</h2>
<p>Per a correlacionar diferents variables. El default és Pearson, normalment es fa servir <strong>Spearman</strong>. Matriu de correlacions</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>cor_data <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> </span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">where</span>(is.numeric))</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(cor_data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">cor_data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">140</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 11%">
<col style="width: 16%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 6%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">gene_1</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">115.60</td>
<td style="text-align: right;">28.32</td>
<td style="text-align: right;">61.3</td>
<td style="text-align: right;">93.70</td>
<td style="text-align: right;">106.7</td>
<td style="text-align: right;">139.25</td>
<td style="text-align: right;">170.7</td>
<td style="text-align: left;">▂▇▃▃▃</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_2</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">119.72</td>
<td style="text-align: right;">30.46</td>
<td style="text-align: right;">67.2</td>
<td style="text-align: right;">97.00</td>
<td style="text-align: right;">111.4</td>
<td style="text-align: right;">141.20</td>
<td style="text-align: right;">192.4</td>
<td style="text-align: left;">▅▇▅▃▂</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gene_3</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">0.93</td>
<td style="text-align: right;">119.48</td>
<td style="text-align: right;">29.84</td>
<td style="text-align: right;">50.7</td>
<td style="text-align: right;">97.30</td>
<td style="text-align: right;">109.0</td>
<td style="text-align: right;">146.48</td>
<td style="text-align: right;">182.1</td>
<td style="text-align: left;">▁▇▅▅▃</td>
</tr>
<tr class="even">
<td style="text-align: left;">gene_4</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">119.15</td>
<td style="text-align: right;">29.24</td>
<td style="text-align: right;">67.0</td>
<td style="text-align: right;">96.05</td>
<td style="text-align: right;">110.2</td>
<td style="text-align: right;">143.40</td>
<td style="text-align: right;">182.6</td>
<td style="text-align: left;">▂▇▃▅▂</td>
</tr>
</tbody>
</table>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(cor_data, <span class="at">method =</span> <span class="st">"spearman"</span>, <span class="at">use=</span><span class="st">"pairwise.complete.obs"</span> )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_1    gene_2    gene_3    gene_4
gene_1 1.0000000 0.7369433 0.7620866 0.7416636
gene_2 0.7369433 1.0000000 0.7440496 0.7148769
gene_3 0.7620866 0.7440496 1.0000000 0.6952780
gene_4 0.7416636 0.7148769 0.6952780 1.0000000</code></pre>
</div>
</div>
<p><code>use="pairwise.complete.obs"</code> d’aquesta manera només elimina NA quan fa aquella comparació, no en la resta)</p>
<p>Per a correlacionar-ne dues del teu interés:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(<span class="at">x=</span>cor_data<span class="sc">$</span>gene_1, <span class="at">y=</span>cor_data<span class="sc">$</span>gene_2, <span class="at">method =</span> <span class="st">"spearman"</span>, <span class="at">use=</span><span class="st">"pairwise.complete.obs"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.7369433</code></pre>
</div>
</div>
<p>Per fer la matriu de scatter-plots</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>( <span class="at">x =</span> cor_data )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Tutorial-Rstudio_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Per triar quines vols observar</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>( <span class="at">formula =</span> <span class="sc">~</span> gene_1 <span class="sc">+</span> gene_2,</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">data =</span> cor_data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Tutorial-Rstudio_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="gràfics" class="level1">
<h1>Gràfics</h1>
<p>Per a fer gràfics R ja et dona opcions per defecte, però per a triar i personalitzar certs aspectes és millor fer-ho amb <code>ggplot</code> (inclòs a <code>tidyverse</code>). Recomano molt tenir la Cheatsheet de <code>ggplot2</code> impresa per anar consultant les ocpions.</p>
<section id="pcr" class="level2">
<h2 class="anchored" data-anchor-id="pcr">PCR</h2>
<p>Per a graficar les dades de PCR, farem un núvol de punts amb línia a la mitjana i barres d’error.</p>
<p>Comencem fent tot el que ja hem explicat</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Carregar els paquets bàsics -------</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor)</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(skimr)</span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Importar dades i preprarar-les-------</span></span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"PCR.xlsx"</span>)) </span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a><span class="co">#crear dataset amb el que treballarem</span></span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> raw_data <span class="sc">%&gt;%</span> </span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clean_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(group, sex, rat, generation), as.factor)) </span>
<span id="cb81-15"><a href="#cb81-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-16"><a href="#cb81-16" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(data) </span>
<span id="cb81-17"><a href="#cb81-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-18"><a href="#cb81-18" aria-hidden="true" tabindex="-1"></a><span class="co">#creem long data</span></span>
<span id="cb81-19"><a href="#cb81-19" aria-hidden="true" tabindex="-1"></a>longdata <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> </span>
<span id="cb81-20"><a href="#cb81-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="st">"gene_1"</span><span class="sc">:</span><span class="st">"gene_4"</span>, <span class="at">names_to=</span><span class="st">"gene"</span>, <span class="at">values_to =</span> <span class="st">"value"</span> ) <span class="sc">%&gt;%</span> </span>
<span id="cb81-21"><a href="#cb81-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb81-22"><a href="#cb81-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> <span class="fu">factor</span>(group, </span>
<span id="cb81-23"><a href="#cb81-23" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>, <span class="st">"S"</span>, <span class="st">"PGS"</span>)), </span>
<span id="cb81-24"><a href="#cb81-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">generation =</span> <span class="fu">factor</span>(generation, </span>
<span id="cb81-25"><a href="#cb81-25" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> <span class="fu">c</span>( <span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>, <span class="st">"CA"</span>)),</span>
<span id="cb81-26"><a href="#cb81-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">gene =</span> <span class="fu">factor</span>(gene,</span>
<span id="cb81-27"><a href="#cb81-27" aria-hidden="true" tabindex="-1"></a>                    <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>)))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Seguidament podem fer un gràfic previ per veure com es comporten els gens en general, sense separar per generació.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>longdata <span class="sc">%&gt;%</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation) <span class="sc">%&gt;%</span> </span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>group, <span class="at">y=</span>value, <span class="at">colour=</span>generation))<span class="sc">+</span> </span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_jitter</span>(<span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.6</span>))<span class="sc">+</span> </span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_discrete</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>, <span class="st">"S"</span>,<span class="st">"PGS"</span>))<span class="sc">+</span> </span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>gene) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Tutorial-Rstudio_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Com podeu veure abans de fer el gràfic hem preparat les dades en la mateixa pipeline, ometent els NA i agrupant per generacions. Seguidament li hem demanat al gràfic:</p>
<ul>
<li><p><code>ggplot(aes(x=group, y=value, colour=generation))+</code> quin és l’eix de les x, quin és l’eix de les y i que determina el color</p></li>
<li><p><code>geom_jitter(position = position_dodge(width = 0.6))+</code> fer un gràfic de punts amb els valors espaiats</p></li>
<li><p><code>scale_x_discrete(limits = c("REF", "P", "G", "S","PGS"))+</code> l’ordre en que apareixen els grups a l’eix de les x</p></li>
<li><p><code>facet_wrap(~gene)</code> indica que ho separi en mini gràfics, un per cada gen. També podriem fer servir <code>facet_grid(~)</code> per a separar-ho de manera més controlada, ho veurem al següent.</p></li>
</ul>
<p>Anem a personalitzar-lo més i afegir les barres d’error i mitjana. Abans de res hem de crear un dataset que contingui la mitjana i error estàndard de cada grup per cada gen dins de cada generació.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute summary stats per a afegir al gràfic-------</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>summary_data <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(generation, gene, group) <span class="sc">%&gt;%</span> </span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean =</span> <span class="fu">mean</span>(value), <span class="co">#calculem mitjana</span></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">se =</span> <span class="fu">sd</span>(value)<span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">n</span>()), <span class="co">#calculem error estàndard</span></span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">.groups =</span> <span class="st">"drop"</span> </span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a><span class="co">#canviar estil del gràfic default</span></span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_classic</span>()) </span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-14"><a href="#cb83-14" aria-hidden="true" tabindex="-1"></a><span class="co">#definim els colors que volem per grup</span></span>
<span id="cb83-15"><a href="#cb83-15" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"REF"</span> <span class="ot">=</span> <span class="st">"#999999"</span>,   <span class="co"># Grey</span></span>
<span id="cb83-16"><a href="#cb83-16" aria-hidden="true" tabindex="-1"></a>            <span class="st">"PGS"</span> <span class="ot">=</span> <span class="st">"#1b9e77"</span>,   <span class="co"># Green 1</span></span>
<span id="cb83-17"><a href="#cb83-17" aria-hidden="true" tabindex="-1"></a>            <span class="st">"S"</span> <span class="ot">=</span> <span class="st">"#66a61e"</span>,   <span class="co"># Green 2</span></span>
<span id="cb83-18"><a href="#cb83-18" aria-hidden="true" tabindex="-1"></a>            <span class="st">"G"</span> <span class="ot">=</span> <span class="st">"#a6d854"</span>,   <span class="co"># Green 3</span></span>
<span id="cb83-19"><a href="#cb83-19" aria-hidden="true" tabindex="-1"></a>            <span class="st">"P"</span> <span class="ot">=</span> <span class="st">"#d9f0a3"</span>)   <span class="co"># Green 4</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Seguidament ja podem fer el gràfic final.</p>
<p>Primer canviem l’ordre en el que apareixeran els gens i el nom del grup de la generació per a que ens surti complet.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co">#reordenar com sortiran els gens com interessi</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>gene_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>)</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>longdata<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">factor</span>(longdata<span class="sc">$</span>gene, <span class="at">levels =</span> gene_order)</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>summary_data<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">factor</span>(summary_data<span class="sc">$</span>gene, <span class="at">levels =</span> gene_order)</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a><span class="co">#canvi nom generation</span></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>longdata<span class="sc">$</span>generation <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>  longdata<span class="sc">$</span>generation,</span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>,<span class="st">"CA"</span>),</span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Dams"</span>,<span class="st">"Day-1 offspring"</span>, <span class="st">"Day-21 offspring"</span>, <span class="st">"Adult offspring"</span>))</span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a>summary_data<span class="sc">$</span>generation <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a>  summary_data<span class="sc">$</span>generation,</span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"Cd1"</span>, <span class="st">"Cd21"</span>,<span class="st">"CA"</span>),</span>
<span id="cb84-15"><a href="#cb84-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Dams"</span>,<span class="st">"Day-1 offspring"</span>, <span class="st">"Day-21 offspring"</span>, <span class="st">"Adult offspring"</span> ))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Ja podem fer el gràfic</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co">#GRÀFIC BO TOTALS------</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>ggplot <span class="ot">&lt;-</span> longdata <span class="sc">%&gt;%</span> <span class="co">#dataset a graficar en format "long"</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span> <span class="co">#no tenir en compte els NA</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>group, <span class="at">y=</span>value, <span class="at">colour=</span>group))<span class="sc">+</span> </span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>()<span class="sc">+</span> <span class="co">#gràfic punts</span></span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>, <span class="st">"S"</span>,<span class="st">"PGS"</span>), </span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">add =</span> <span class="dv">1</span>))<span class="sc">+</span> <span class="co">#afegir espai entre grups</span></span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(generation<span class="sc">~</span>gene, </span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">labeller =</span> <span class="fu">labeller</span>(<span class="at">gene =</span> gene_order))<span class="sc">+</span> <span class="co">#ordre </span></span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">100</span>, <span class="at">color =</span> <span class="st">"black"</span>,<span class="at">linetype =</span> <span class="st">"dotted"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>)<span class="sc">+</span></span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="at">data =</span> summary_data, </span>
<span id="cb85-12"><a href="#cb85-12" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> mean, <span class="at">ymin =</span> mean <span class="sc">-</span> se, <span class="at">ymax =</span> mean <span class="sc">+</span> se, <span class="at">width =</span> <span class="fl">0.25</span>))<span class="sc">+</span> <span class="co">#afegir barres d'error</span></span>
<span id="cb85-13"><a href="#cb85-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>( <span class="co">#afegir línia mitjana</span></span>
<span id="cb85-14"><a href="#cb85-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> summary_data,</span>
<span id="cb85-15"><a href="#cb85-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> group,</span>
<span id="cb85-16"><a href="#cb85-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">ymin =</span> mean,</span>
<span id="cb85-17"><a href="#cb85-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">ymax =</span> mean,</span>
<span id="cb85-18"><a href="#cb85-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">colour =</span> group),</span>
<span id="cb85-19"><a href="#cb85-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">width =</span> <span class="fl">0.8</span>,</span>
<span id="cb85-20"><a href="#cb85-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">inherit.aes =</span> <span class="cn">FALSE</span>,</span>
<span id="cb85-21"><a href="#cb85-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="dv">1</span>)<span class="sc">+</span> </span>
<span id="cb85-22"><a href="#cb85-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values=</span>colors, </span>
<span id="cb85-23"><a href="#cb85-23" aria-hidden="true" tabindex="-1"></a>                      <span class="at">breaks =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>, <span class="st">"S"</span>, <span class="st">"PGS"</span>))<span class="sc">+</span> </span>
<span id="cb85-24"><a href="#cb85-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span> <span class="st">"PCR small intestine"</span>, </span>
<span id="cb85-25"><a href="#cb85-25" aria-hidden="true" tabindex="-1"></a>       <span class="at">x=</span> <span class="st">"Group"</span>, </span>
<span id="cb85-26"><a href="#cb85-26" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span> <span class="st">"Relative mRNA levels (%)"</span>,</span>
<span id="cb85-27"><a href="#cb85-27" aria-hidden="true" tabindex="-1"></a>       <span class="at">colour=</span> <span class="st">"Group"</span>)<span class="sc">+</span>  </span>
<span id="cb85-28"><a href="#cb85-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>), </span>
<span id="cb85-29"><a href="#cb85-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>),</span>
<span id="cb85-30"><a href="#cb85-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>, <span class="at">hjust =</span> <span class="fl">0.75</span>),</span>
<span id="cb85-31"><a href="#cb85-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb85-32"><a href="#cb85-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.spacing =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>)) <span class="co">#espai entre els mini gràfics</span></span>
<span id="cb85-33"><a href="#cb85-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-34"><a href="#cb85-34" aria-hidden="true" tabindex="-1"></a>ggplot</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Tutorial-Rstudio_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><p><code>geom_hline(yintercept = 100, color = "black",linetype = "dotted", size = 0.5)+</code> Afegim una línia de punts al 100</p></li>
<li><p><code>geom_errorbar(data = summary_data, aes(x = group_name, y = mean, ymin = mean - se, ymax = mean + se, width = 0.25))+</code> Afegim les barres d’error</p></li>
<li><p><code>geom_errorbar(data = summary_data, aes( x = group, ymin = mean, ymax = mean, colour = group), width = 0.8, inherit.aes = FALSE, linewidth = 1 )+</code> Afegim línia a la mitjana (no hi ha manera de fer-ho directe).</p></li>
<li><p><code>labs(title= "PCR small intestine", gràfic x= "Group",  eix x y= "Relative mRNA levels (%)", colour= "Group")+</code> Et permet canviar el nom dels eixos i títol i nom de la llegenda, en aquest cas ens la fa del color però també la podria fer per colour, fill, shape, linetype, size…</p></li>
</ul>
<p>Un altre exemple seria triar només una generació, on només canvia el següent</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>summary_data_d1 <span class="ot">&lt;-</span><span class="fu">filter</span>(summary_data, generation <span class="sc">==</span> <span class="st">"Day-1 offspring"</span>) <span class="co">#filtrem summary_data</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>longdata <span class="sc">%&gt;%</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(generation<span class="sc">==</span><span class="st">"Day-1 offspring"</span>) <span class="sc">%&gt;%</span> <span class="co">#afegim filtre!!</span></span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>group, <span class="at">y=</span>value, <span class="at">colour=</span>group))<span class="sc">+</span></span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>()<span class="sc">+</span></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>,<span class="st">"PGS"</span>),</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">add =</span> <span class="dv">1</span>))<span class="sc">+</span></span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="sc">~</span>gene,    <span class="co">#només separem per gen</span></span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">labeller =</span> <span class="fu">labeller</span>(<span class="at">gene =</span> gene_order))<span class="sc">+</span></span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">100</span>, <span class="at">color =</span> <span class="st">"black"</span>,<span class="at">linetype =</span> <span class="st">"dotted"</span>, <span class="at">size =</span> <span class="fl">0.5</span>)<span class="sc">+</span> </span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="at">data =</span> summary_data_d1, <span class="co">#canvi dataset</span></span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a>                <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> mean, <span class="at">ymin =</span> mean <span class="sc">-</span> se, <span class="at">ymax =</span> mean <span class="sc">+</span> se, <span class="at">width =</span> <span class="fl">0.25</span>))<span class="sc">+</span></span>
<span id="cb86-15"><a href="#cb86-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="at">data =</span> summary_data_d1, <span class="co">#canvi dataset</span></span>
<span id="cb86-16"><a href="#cb86-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> group,</span>
<span id="cb86-17"><a href="#cb86-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">ymin =</span> mean,</span>
<span id="cb86-18"><a href="#cb86-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">ymax =</span> mean,</span>
<span id="cb86-19"><a href="#cb86-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> group),</span>
<span id="cb86-20"><a href="#cb86-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">width =</span> <span class="fl">0.8</span>,</span>
<span id="cb86-21"><a href="#cb86-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">inherit.aes =</span> <span class="cn">FALSE</span>,</span>
<span id="cb86-22"><a href="#cb86-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="dv">1</span>)<span class="sc">+</span> </span>
<span id="cb86-23"><a href="#cb86-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values=</span>colors,</span>
<span id="cb86-24"><a href="#cb86-24" aria-hidden="true" tabindex="-1"></a>                <span class="at">breaks =</span> <span class="fu">c</span>(<span class="st">"REF"</span>, <span class="st">"P"</span>, <span class="st">"G"</span>, <span class="st">"PGS"</span>))<span class="sc">+</span></span>
<span id="cb86-25"><a href="#cb86-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span> <span class="st">"PCR small intestine Day-1 Offspring"</span>, </span>
<span id="cb86-26"><a href="#cb86-26" aria-hidden="true" tabindex="-1"></a>       <span class="at">x=</span> <span class="st">"Group"</span>,</span>
<span id="cb86-27"><a href="#cb86-27" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span> <span class="st">"Relative mRNA levels (%)"</span>,</span>
<span id="cb86-28"><a href="#cb86-28" aria-hidden="true" tabindex="-1"></a>       <span class="at">colour=</span> <span class="st">"Group"</span>)<span class="sc">+</span> </span>
<span id="cb86-29"><a href="#cb86-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>), </span>
<span id="cb86-30"><a href="#cb86-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>),</span>
<span id="cb86-31"><a href="#cb86-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb86-32"><a href="#cb86-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.spacing =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Tutorial-Rstudio_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>I finalment ho guardem, si veiem que surt algo tallat o massa apretat, podem fer la width o height més gran. Aquesta instrucció ens guardarà l’últim plot.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">here</span>(<span class="st">"output"</span>,<span class="st">"nom fitxer.png"</span>), <span class="at">width =</span> <span class="dv">18</span>, <span class="at">height =</span> <span class="dv">12</span>, <span class="at">dpi =</span> <span class="dv">300</span>) </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Falta veure com afegiriem la significància.</p>
</section>
<section id="heatmap-log2fc" class="level2">
<h2 class="anchored" data-anchor-id="heatmap-log2fc">Heatmap log<sub>2</sub>(FC)</h2>
<p>Per a fer un heatmap amb augment o disminució de concentracions respecte al grup REF amb la significància, la qual tenim en una altre taula d’excel.</p>
<p>On ho vaig aprendre: <a href="https://youtu.be/369PHkv1fPg?feature=shared" class="uri">https://youtu.be/369PHkv1fPg?feature=shared</a></p>
<p>En el cas de que tinguem els valors per cada observació, primer haurem de calcular el log<sub>2</sub>(FC) de les dades.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Carregar els paquets bàsics -------</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor)</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(skimr)</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a><span class="co">#paquet extra: </span></span>
<span id="cb89-9"><a href="#cb89-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb89-10"><a href="#cb89-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-11"><a href="#cb89-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Importar dades i preprarar-les-------</span></span>
<span id="cb89-12"><a href="#cb89-12" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"PCR.xlsx"</span>)) </span>
<span id="cb89-13"><a href="#cb89-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-14"><a href="#cb89-14" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> raw_data <span class="sc">%&gt;%</span> </span>
<span id="cb89-15"><a href="#cb89-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clean_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb89-16"><a href="#cb89-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(group, sex, rat, generation), as.factor)) </span>
<span id="cb89-17"><a href="#cb89-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-18"><a href="#cb89-18" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Convertim les dades</p>
<p>A partir d’aquí s’ha de revisar</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>data_log2<span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span> </span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">%&gt;%</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">where</span>(is.numeric),</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>    <span class="sc">~</span> <span class="fu">ifelse</span>(<span class="fu">is.na</span>(.), <span class="fu">mean</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>), .)</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span> </span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(</span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a>      <span class="dv">3</span><span class="sc">:</span><span class="fu">last_col</span>(),</span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>      <span class="sc">~</span> .x <span class="sc">/</span> <span class="fu">mean</span>(.x [group <span class="sc">==</span> <span class="st">"REF"</span>], <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>    )) <span class="sc">%&gt;%</span> <span class="co">#canviem els NA per la mitjana d'aquell metabolit dins del grup</span></span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="dv">3</span><span class="sc">:</span><span class="fu">last_col</span>(), log2)) <span class="co">#aplicar logaritme</span></span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a>data_HM <span class="ot">&lt;-</span> data_log2 <span class="sc">%&gt;%</span></span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>group) <span class="sc">%&gt;%</span>              <span class="co"># 1️⃣ eliminar columna group</span></span>
<span id="cb90-16"><a href="#cb90-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"rat"</span>) <span class="sc">%&gt;%</span>   <span class="co"># 2️⃣ rat → rownames</span></span>
<span id="cb90-17"><a href="#cb90-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span>                         <span class="co"># 3️⃣ transposar</span></span>
<span id="cb90-18"><a href="#cb90-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb90-19"><a href="#cb90-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-20"><a href="#cb90-20" aria-hidden="true" tabindex="-1"></a>mat_data <span class="ot">&lt;-</span> <span class="fu">data.matrix</span>(data_HM) <span class="co">#fem una matriu</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Espero que us hagi servit!</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>